---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# txtools

<!-- badges: start -->
<!-- badges: end -->

The goal of txtools is to ...

## Installation

Install txtools dependencies with the following code

```{r, eval = F}
# CRAN packages
install.packages(c("magrittr", "stringr", "devtools"))
# Bioconductor packages
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install(c("GenomicRanges", "plyranges", "Rsamtools", "GenomicAlignments"))
```

Install the development version from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("AngelCampos/testRepo")
BiocManager::install("AngelCampos/txtools")
```
## Demo

This is a basic example which shows you how to solve a common problem:

```{r example}
library(txtools)
bamFile <- system.file("extdata", "example_hg19.bam", package = "txtools")
reads <- load_pairedEnd_bam(bamFile)
```



```{r cars}
bedFile <- system.file("extdata", "twoUCSCgenes_hg19.bed", package = "txtools")
geneAnnot <- plyranges::read_bed(bedFile)
txReads <- tx_PEreads(reads, geneAnnot)

```

You'll still need to render `README.Rmd` regularly, to keep `README.md` up-to-date.

You can also embed plots, for example:

```{r pressure, echo = FALSE}
plot(pressure)
```

In that case, don't forget to commit and push the resulting figure files, so they display on GitHub!
