---
output: github_document
editor_options: 
  chunk_output_type: console
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# txtools

<!-- badges: start -->
<!-- badges: end -->

The goal of txtools is to ...

## Installation

Install txtools dependencies with the following code

```{r, eval = F}
# CRAN packages
install.packages(c("magrittr", "stringr", "devtools"))
# Bioconductor packages
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install(c("GenomicRanges", "plyranges", "Rsamtools", "GenomicAlignments"))
```

Install the development version from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("AngelCampos/testRepo")
BiocManager::install("AngelCampos/txtools")
```
## Demo

This is a basic example which shows you how to solve a common problem:

```{r example}
library(txtools)
# This example files are installed along txtools
bamFile <- system.file("extdata", "example_hg19.bam", package = "txtools")
bedFile <- system.file("extdata", "twoUCSCgenes_hg19.bed", package = "txtools")

reads <- load_pairedEnd_bam(bamFile, loadSeq = T, verbose = F)
geneAnnot <- plyranges::read_bed(bedFile) # plyranges read_bed function
txReads <- tx_PEreads(reads, geneAnnot, withSeq = T)
txReads$uc003lam.1
```


The main object 


```{r example2}
library(magrittr) # %>% Pipe operator 

resTab <- hlp_cbind3Tabs(tx_genCoorTab(txReads, geneAnnot),
                         tx_covTab(txReads),
                         tx_nucFreqTab(txReads, simplify_IUPAC = "splitForceInt"))

resTab$uc003lam.1[,..IUPAC_code_simpl] %>% data.frame() %>% t %>% 
    barplot(col = c(RColorBrewer::brewer.pal(4, "Set1"), 
                    "black", "white", "gray"), border = NA, main= "uc003lam.1")
```

