---
title: "txtools_core"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{txtools_core}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",  
  fig.width = 6, 
  fig.height = 4
)
```

# txtools. The full tour.


The aim of txtools is to provide a versatile framework to analyze transcriptomic
data in a quick and simple manner. 

To acomplish this we propose the processing of data in two stages, and the 
use of two data structures which is the output at each step:

1. Transfer read-alignments from a **genomic** to  a **transcriptomic**
coordinates space.
2. Summarize transcriptomic-alignments into table structures.

## Loading the data

The first step of the txtools processing pipeline is to transfer read-alignments
from a genomic coordinate system to a transcriptomic coordinate system. This 
step can be skipped if the reads where aligned to a transcriptomic reference.

Note: *This step can be circumvented by aligning to the transcriptome, but in most* 
*cases it is necessary and desirable to keep track of both the genomic location*
*of transcripts and the read-alignments data.*

To use txtools it is REQUIRED to have a **Gene Annotation** file in the
form of a BED-6/12 file. Importantly, a BED-12 file contains the information 
to construct the exon structure, while using a BED6 assumes that the gene
consists of a single exon block.

Additionally it is recommended to have a **Reference Genome** sequence file
in the form of a FASTA file, in which each entry corresponds to each chromosome.

Using txtools we can seamlessly load RNA-seq data into the R environment
with the `tx_load_bam()` function. For this example we will use a paired-end
BAM file and we will require the function to load the sequence data.

```{r}
library(txtools)
library(pasillaBamSubset)

# Loading gene-annotation in BED-12 format.
pathToBED <- tx_dm3_geneAnnot()
dm3_geneAnnot <- tx_load_bed(pathToBED)

# Loading genome
fastaFile <- pasillaBamSubset::dm3_chr4()
dm3_genome <- tx_load_genome(fastaFile)

# Loading RNA-seq data (paired-end)
pathToBAM <- pasillaBamSubset::untreated3_chr4()
bam_PE <- tx_load_bam(file = pathToBAM, pairedEnd = TRUE, loadSeq = TRUE)
summary(bam_PE)
```

## From Genomic to Transcriptomic

### tx_reads()

Now to convert the genomic-alignments into their transcriptomic versions we
use the `tx_reads()` function, and require it to keep the sequence data 
setting the `withSeq` argument to 'TRUE', this will allow us to later summarize
nucleotide frequency. If nucleotide frequency is not required in downstream 
analysis setting `withSeq` to 'FALSE' will expedite the processing. 

Importantly, `tx_reads()` can use multiple cores to accelerate processing,
just set the `nCores` argument to the ammount of cores you want to use.
In practice a number of cores between 6-8 is enough, but this may depend on 
your infrastructure.

```{r}
txreads_PE <- tx_reads(reads = bam_PE, geneAnnot = dm3_geneAnnot, 
                       withSeq = TRUE, minReads = 1, nCores = 1)
```

The resulting object is a `CompressedGRangesList`, a list of GRanges,
grouped by their corresponding gene. `GRanges` is a class from the
`GenomicRanges` package. Therefore they can be manipulated using this 
package. At the time of writing, txtools does not focus on the use of `GRanges`.
nevertheless provides some functions to manipulate them before summarizing them
into data.tables.

### txtools function for GRangesLists

**tx_counts**

A function that simply counts the num

```{r}
tx_counts(txreads_PE)
```


**tx_sample_GRList()**

A function that samples GRanges from a GRangesList, useful for simulation 
of lower coverages in samples.

```{r}
txreads_PE_half <- tx_sample_GRList(txreads_PE, 0.5)
tx_counts(txreads_PE_half)
```

**tx_filter_maxWidth()**

A function that filters out Granges longer than a certain width. Useful to
filter spurious alignments that are not possible due to RNA insert length
selection.

```{r, fig.height = 2, fig.width= 6, fig.show="hold"}
txreads_PE_filtered <- tx_filter_maxWidth(txreads_PE, 600)

boxplot(unlist(GenomicRanges::width(txreads_PE)), xlab = "Insert Width",
        horizontal = T, main = "Not filtered alignments") 
boxplot(unlist(GenomicRanges::width(txreads_PE_filtered)), xlab = "Insert Width",
        horizontal = T, main = "Filtered alignments") 
```

## Summarizing transcriptomic GRanges

### tx_makeDT_* functions

Due to GRanges contrains it is not possible to keep summarized metrics into the 
same structure as GRanges or GRangesList objects. For example, the coverage
of a certain gene pertains to an specific range of coordinates and cannot be 
assigned to an element in particular. This is the reason we opt to create
a summarized table that holds the relevant information from the reads in a way
that is easier to interpret and further analyze.

To summarize GRangesList txtools contains three functions which do the job:

* **tx_makeDT_coverage()**
* **tx_makeDT_nucFreq()**
* **tx_makeDT_covNucFreq()**


```{r}
DT_cov <- tx_makeDT_covNucFreq(txreads_PE_filtered, geneAnnot = dm3_geneAnnot,
                     dm3_genome, nCores = 1)
unique(DT_cov$gene)
barplot(DT_cov[gene == "NM_079879",]$cov)
```

2. Summarize the read-alignments corresponding to each gene model into 
a table structure, that keeps per nucleotide information. 

The two main functions to process *genomic alignments* data
into **transcriptomic** data are `tx_reads()` and `tx_makeDT_*()` family 
functions.

To fully understand what happens under the hood using txtools to process your 
data we will use toy examples of **genomic alignments** and **gene annotation**,
the main inputs to process the data into transcriptomic data.
