---
title: "txtools_tutorial"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{txtools_tutorial}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

This will be a more in depth tutorial showing all of the main functions.

We will check how to:

1. Load your data RNA-seq data and genomic reference files to R.
2. Process the data into transcriptomic space and filter 
3. Summarize reads into data.tables
4. Use tx_add_* functions to calculate metrics into data.tables
5. Plot interesting regions using the tx_plot_* functions

```{r}

```

