---
title: "txtools_tutorial"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{txtools_tutorial}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

This will be a more in depth tutorial on how to use txtools.

We will check how to:

1. Load your data RNA-seq data and genomic reference files to R.
2. Process the data into transcriptomic space and filter 
3. Summarize reads into data.tables
4. Use tx_add_* functions to calculate metrics into data.tables
5. Plot interesting regions using the tx_plot_* functions


```{r, fig.width = 6, fig.height= 4}
# Load package
library(txtools)
library(pasillaBamSubset)

# Loading gene Annotation
BEDfile <- system.file("extdata", "dm3_chr4_RefSeqGenes_UCSC_top10genes.bed", 
                       package = "txtools")
dm3_geneAnnot <- tx_load_bed(BEDfile)
# Loading genome
fastaFile <- pasillaBamSubset::dm3_chr4()
dm3_genome <- tx_load_genome(fastaFile)
# Loading BAM files
singleEndBamFile <- pasillaBamSubset::untreated1_chr4()

bam_SE <- tx_load_bam(file = singleEndBamFile, pairedEnd = F, loadSeq = T, verbose = F)
reads_SE <- tx_reads(reads = bam_SE, 
                     geneAnnot = dm3_geneAnnot, 
                     withSeq = T, 
                     nCores = 10, 
                     minReads = 1)
DT <- tx_makeDT_covNucFreq(reads_SE, geneAnnot = dm3_geneAnnot, genome = dm3_genome)
DT <- tx_add_diffNucToRefRatio(DT, addDiffandTotalCols = TRUE)

DT[which(DT$diffToRefRatio > 0.8 & cov > 70),]

tx_plot_nucFreq(DT, gene = "NM_079901", txRange = window_around(2171, 10), makePlotly = T)
    
```



