<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Changelog • txtools</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Changelog"><meta property="og:image" content="https://angelcampos.github.io/txtools/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">txtools</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.5</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/txtools.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-news">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Changelog</h1>
      
    </div>

    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.1.5" id="txtools-015">txtools 0.1.5<a class="anchor" aria-label="anchor" href="#txtools-015"></a></h2>
<ul><li>Added the <code>loadSecondaryAligns</code> argument to <code><a href="../reference/tx_load_bam.html">tx_load_bam()</a></code> so that secondary alignments can be easily discarded by setting it to FALSE. The same effect can be accomplished by inputing a scanFlag built with the <code><a href="https://rdrr.io/pkg/Rsamtools/man/ScanBamParam-class.html" class="external-link">Rsamtools::scanBamFlag()</a></code> helper, but this way it is a more streamlined feature.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.1.4" id="txtools-014">txtools 0.1.4<a class="anchor" aria-label="anchor" href="#txtools-014"></a></h2>
<ul><li>Added tx_combineTxReads() to combine reads processed by tx_reads().</li>
<li>Setup pkgdown <a href="https://angelcampos.github.io/txtools">website</a>.</li>
<li>Bug fix: Loading Bed-6 gene annotations broke due to a dependency update.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.1.2" id="txtools-012">txtools 0.1.2<a class="anchor" aria-label="anchor" href="#txtools-012"></a></h2>
<ul><li>Bug fix: stringr::str_sub() update broke functions.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.1.1" id="txtools-011">txtools 0.1.1<a class="anchor" aria-label="anchor" href="#txtools-011"></a></h2>
<ul><li>Bug fix: Error processing single-end reads</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.1.0" id="txtools-010">txtools 0.1.0<a class="anchor" aria-label="anchor" href="#txtools-010"></a></h2>
<ul><li>Major:
<ul><li>Added ignore.strand functionality to load reads regardless if the RNA-seq library is strand-aware.</li>
<li>New plotting function tx_plot_numeric(): Plots numeric variables along the transcriptomic space.</li>
<li>tx_metagene_<em>() functions have a normalization argument which yields the are under the curve to be proportional to 1</em>(transcriptomic window length).</li>
</ul></li>
<li>Plus minor fixes.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.8" id="txtools-008">txtools 0.0.8<a class="anchor" aria-label="anchor" href="#txtools-008"></a></h2>
<ul><li>Fixed loading paired end BAM files when strand sign is defined by read2 (“last”).</li>
<li>Bugs fixed</li>
<li>Minor improvements:
<ul><li>Loading genome name as first word separated by spaces</li>
<li>Edited main vignette tutorial “txtools”</li>
<li>Added bam2txDT.R script in inst/ dir</li>
</ul></li>
<li>Patch 0.0.8.1:
<ul><li>Added “splicingSite” option to tx_get_metageAtCDS() and tx_plot_metageneAtCDS().</li>
</ul></li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.7" id="txtools-007">txtools 0.0.7<a class="anchor" aria-label="anchor" href="#txtools-007"></a></h2>
<ul><li>General Update</li>
<li>Added vignette “txtools_user_guide”</li>
<li>Changed tx_add_diffNucToRef() for tx_add_misincCount()</li>
<li>Changed tx_add_diffNucToRefRatio() for tx_add_misincRate()</li>
<li>Fix minor bugs</li>
<li>
<strong>0.0.7.4 patch</strong>: Fixed a bug that incorrectly included alignments with N operations in their CIGAR strings (gaps), that didn’t match the gene structure.</li>
<li>Removed plotly conversion in tx_plot_*() functions. As can be easily done by the user with plotly::ggplotly()</li>
<li>
<strong>0.0.7.8 patch</strong>: Fixed tx_orderDT() so that order is not affected by levels in the ‘gene’ factor; accordingly fix tx_unifyTxDTL() so DTs have the same txcoor order. Changing name of function tx_add_misincorpRateNucSpec() in favor of tx_add_misincRateNucSpec(), so that it conforms to other function names, former can be used as alias but has been deprecated.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.6" id="txtools-006">txtools 0.0.6<a class="anchor" aria-label="anchor" href="#txtools-006"></a></h2>
<ul><li>Fixed a bug detected when loading BAM files with mappings with empty sequence field</li>
<li>0.0.6.2 patch: Fixed bug related with spurious paired-end alignments, R1 and R2 order is wrong.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.5" id="txtools-005">txtools 0.0.5<a class="anchor" aria-label="anchor" href="#txtools-005"></a></h2>
<ul><li>Update Bioconductor remotes.</li>
<li>Adding tx_core_mc() which is deprecated as alias of tx_reads().</li>
<li>Fix bug in tx_complete_DT() which added refSeq before necessary.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.4" id="txtools-004">txtools 0.0.4<a class="anchor" aria-label="anchor" href="#txtools-004"></a></h2>
<ul><li>Uni and Multi-core functions merged into unique functions capable of using multi-cores with the parallel package.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.3" id="txtools-003">txtools 0.0.3<a class="anchor" aria-label="anchor" href="#txtools-003"></a></h2>
<ul><li>Bug fixed when using BED6 annotation</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.2" id="txtools-002">txtools 0.0.2<a class="anchor" aria-label="anchor" href="#txtools-002"></a></h2>
<ul><li>Updating plotting function tx_plot_staEndCov() to remove coverage counts.</li>
<li>Other minor changes</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.1" id="txtools-001">txtools 0.0.1<a class="anchor" aria-label="anchor" href="#txtools-001"></a></h2>
<ul><li>Initial draft of package</li>
</ul></div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Miguel Angel Garcia-Campos.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

