<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="txtools is an R package that enables the processing, analysis, and
    visualization of RNA-seq data at the nucleotide-level resolution, seamlessly 
    integrating alignments to the genome with transcriptomic representation.">
<title>An R package facilitating analysis of RNA modifications, structures, and interactions • txtools</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="An R package facilitating analysis of RNA modifications, structures, and interactions">
<meta property="og:description" content="txtools is an R package that enables the processing, analysis, and
    visualization of RNA-seq data at the nucleotide-level resolution, seamlessly 
    integrating alignments to the genome with transcriptomic representation.">
<meta property="og:image" content="https://angelcampos.github.io/txtools/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">txtools</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.2</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="articles/txtools.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header">
<img src="logo.png" class="logo" alt=""><h1 id="txtools-">txtools 
<a class="anchor" aria-label="anchor" href="#txtools-"></a>
</h1>
</div>
<!-- badges: start -->

<div class="section level2">
<h2 id="description">Description<a class="anchor" aria-label="anchor" href="#description"></a>
</h2>
<p><strong>txtools</strong> is a package that processes RNA-seq reads alignments into transcriptomic-oriented tables. Enabling a quick and simplified analysis, to closely inspect summarized RNA-seq data per transcript, at nucleotide resolution, i.e. coverage, read-starts, read-ends, deletions, and nucleotide frequency. Attractive plotting is also readily available to visualize data.</p>
<p>The main processing pipeline of txtools consists of 1) converting genomic alignments into transcriptomic space and merging paired-end reads using <code><a href="reference/tx_reads.html">tx_reads()</a></code> and 2) summarize counts for ‘readouts’ (coverage, read-start, read-ends, nucleotide frequency, deletions) along the transcriptome using a function of the <code>tx_makeDT_</code> family.</p>
<figure><img src="reference/figures/readme_1.png" alt="Figure 1. txtools’ main processing pipeline"><figcaption aria-hidden="true">
Figure 1. txtools’ main processing pipeline
</figcaption></figure><p>Of note, coverage calculation in paired-end reads using txtools considers the whole range of the paired alignment. From the start of read1 to the end of read2, including the insert which is not directly sequenced (Figure 2)</p>
<figure><img src="reference/figures/readme_2.png" style="width:60.0%" alt="Figure 2. Paired-end reads coverage computation"><figcaption aria-hidden="true">
Figure 2. Paired-end reads coverage computation
</figcaption></figure>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>You can install the development version from <a href="https://github.com/AngelCampos/txtools" class="external-link">GitHub</a> typing in the following commands in the R console:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"remotes"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"remotes"</span><span class="op">)</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"AngelCampos/txtools"</span>, build_vignettes <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="quick-example">Quick example<a class="anchor" aria-label="anchor" href="#quick-example"></a>
</h2>
<p>In this small example we use the ‘Pasilla’ experiment data (from its own package) which contains a BAM file for the paired-end alignments of a <em>D. melanogaster</em> RNA-seq experiment on chromosome 4, along with a FASTA file comprising the genome sequence for the same chromosome.</p>
<p>Using txtools we can load the <strong>genome</strong> (FASTA), the <strong>gene annotation</strong> (BED-12), and the <strong>RNA-seq reads alignment</strong> (BAM) files into R with ease.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://angelcampos.github.io/txtools/">txtools</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">pasillaBamSubset</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Getting paths to files</span></span>
<span><span class="va">BED_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/tx_dm3_geneAnnot.html">tx_dm3_geneAnnot</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">FASTA_file</span> <span class="op">&lt;-</span> <span class="fu">pasillaBamSubset</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/pasillaBamSubset/man/pasillaBamSubset-package.html" class="external-link">dm3_chr4</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># Data from pasillaBamSubset pkg</span></span>
<span><span class="va">PE_BAM_file</span> <span class="op">&lt;-</span> <span class="fu">pasillaBamSubset</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/pasillaBamSubset/man/pasillaBamSubset-package.html" class="external-link">untreated3_chr4</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># Data from pasillaBamSubset pkg</span></span>
<span></span>
<span><span class="co"># Loading D. melanogaster gene annotation, genome, and paired-end bam alignments</span></span>
<span><span class="va">dm3_geneAnnot</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/tx_load_bed.html">tx_load_bed</a></span><span class="op">(</span><span class="va">BED_file</span><span class="op">)</span></span>
<span><span class="va">dm3_genome</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/tx_load_genome.html">tx_load_genome</a></span><span class="op">(</span><span class="va">FASTA_file</span><span class="op">)</span></span>
<span><span class="va">dm3_PEreads</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/tx_load_bam.html">tx_load_bam</a></span><span class="op">(</span>file <span class="op">=</span> <span class="va">PE_BAM_file</span>, pairedEnd <span class="op">=</span> <span class="cn">TRUE</span>, loadSeq <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>First, we process the alignments to their transcriptomic versions using the <code><a href="reference/tx_reads.html">tx_reads()</a></code> function.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">reads_SE</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/tx_reads.html">tx_reads</a></span><span class="op">(</span>reads <span class="op">=</span> <span class="va">dm3_PEreads</span>, </span>
<span>                     geneAnnot <span class="op">=</span> <span class="va">dm3_geneAnnot</span>, </span>
<span>                     withSeq <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                     nCores <span class="op">=</span> <span class="fl">1</span>, </span>
<span>                     minReads <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in rmAmbigStrandAligns(reads): Removing 63 alignments with ambiguous strand (*)</span></span>
<span><span class="co">#&gt; Processing 75346 reads, using 10 gene models. </span></span>
<span><span class="co">#&gt; 10373 alignments overlap 10 gene models </span></span>
<span><span class="co">#&gt; Assigning alignments to gene model... </span></span>
<span><span class="co">#&gt; Processing sequences. This may take several minutes depending on geneAnnot size ... </span></span>
<span><span class="co">#&gt; Output contains: 7176 unique alignments in 10 gene models</span></span>
<span><span class="co">#&gt; Warning in tx_reads(reads = dm3_PEreads, geneAnnot = dm3_geneAnnot, withSeq =</span></span>
<span><span class="co">#&gt; TRUE, : Some alignments were not assigned to any gene, you can retrieve them</span></span>
<span><span class="co">#&gt; using the tx_get_unassignedAlignments() function.</span></span></code></pre></div>
<p>Then we just need to summarize the alignments into a txDT. In this case using the <code><a href="reference/tx_makeDT_covNucFreq.html">tx_makeDT_covNucFreq()</a></code> function outputs a table with all the base metrics, including read coverage (‘cov’ column), and nucleotide frequency (A,C,T,G columns).</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DT</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/tx_makeDT_covNucFreq.html">tx_makeDT_covNucFreq</a></span><span class="op">(</span><span class="va">reads_SE</span>, geneAnnot <span class="op">=</span> <span class="va">dm3_geneAnnot</span>, genome <span class="op">=</span> <span class="va">dm3_genome</span><span class="op">)</span></span></code></pre></div>
<p>The resulting txDT comprise all summarized information from the RNA-seq reads aligned to the genome and contained within the genes in the gene annotation (this example consists of only the top 10 expressed genes). For more information on the columns of DT consult the <code><a href="reference/tx_makeDT_covNucFreq.html">tx_makeDT_covNucFreq()</a></code> documentation.</p>
<p>To extend the base metrics that the <code>tx_makeDT_*()</code> functions provide txtools provides the <code>tx_add_*()</code> functions family. One example of such functions is <code><a href="reference/tx_add_diffNucToRefRatio.html">tx_add_diffNucToRefRatio()</a></code> which calculates the ratio of nucleotide counts different to the reference sequence. Using these metric we can easily spot locations in which RNA transcripts sequence is different from that of the reference sequence.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DT</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/tx_add_misincRate.html">tx_add_misincRate</a></span><span class="op">(</span><span class="va">DT</span>, addCounts <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">DT</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">misincRate</span> <span class="op">&gt;</span> <span class="fl">0.5</span> <span class="op">&amp;</span> <span class="va">nucTotal</span> <span class="op">&gt;</span> <span class="fl">40</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="co">#&gt;     chr gencoor strand      gene txcoor refSeq cov start_5p end_3p A C  G T N -</span></span>
<span><span class="co">#&gt; 1: chr4  939355      - NM_079901   3803      A  90        0      0 0 0 46 0 0 0</span></span>
<span><span class="co">#&gt;     . misincCount nucTotal misincRate</span></span>
<span><span class="co">#&gt; 1: 44          46       46          1</span></span></code></pre></div>
<p>Finally, using the <code><a href="reference/tx_plot_nucFreq.html">tx_plot_nucFreq()</a></code> function we can visualize that data in the DT at an specific location.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/tx_plot_nucFreq.html">tx_plot_nucFreq</a></span><span class="op">(</span><span class="va">DT</span>, gene <span class="op">=</span> <span class="st">"NM_079901"</span>, txRange <span class="op">=</span> <span class="fu"><a href="reference/window_around.html">window_around</a></span><span class="op">(</span><span class="fl">3803</span>, <span class="fl">15</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-plotNucFreq-1.png"><!-- --></p>
</div>
<div class="section level2">
<h2 id="further-documentation">Further documentation<a class="anchor" aria-label="anchor" href="#further-documentation"></a>
</h2>
<ul>
<li>The <em>txtools</em> user guide is available as a vignette typing <code><a href="articles/txtools.html">vignette("txtools")</a></code> at the R console.</li>
<li>The use cases code is available at its own repo in: <a href="https://github.com/AngelCampos/txtools_uc" class="external-link uri">https://github.com/AngelCampos/txtools_uc</a>.</li>
</ul>
</div>
<div class="section level2">
<h2 id="current-limitations">Current limitations:<a class="anchor" aria-label="anchor" href="#current-limitations"></a>
</h2>
<ul>
<li><p>Insertions: txtools is not able to deal with insertions. This is mainly because insertions are not part of the original trasncriptomic nor genomic reference space as they would alter the length of the gene model. This could be an added feature in future versions but is not a priority.</p></li>
<li><p>Potentially long processing times: Loading big BAM files into R commonly requires a lot of time, having this in mind txtools provides a progress bar to keep users informed about the loading status. Most importantly, depending on the ammount of both loaded reads and the size of the <em>Gene Annotation</em> tx_reads() processing time can take several minutes. A solution to this issue is the use of multi-threadding which has been incorporated into tx_reads() and other functions, but such functionality is only available for UNIX and MAC OS.</p></li>
</ul>
</div>
<div class="section level2">
<h2 id="additional-notes">Additional notes:<a class="anchor" aria-label="anchor" href="#additional-notes"></a>
</h2>
<ul>
<li>As many R packages meant for high-throughput data analysis and manipulation, using <strong><em>txtools</em></strong> may require high ammounts of RAM memory, depending mainly on the size of BAM files being processed.</li>
</ul>
</div>
</div>
  </main><aside class="col-md-3"><div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing txtools</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Miguel Angel Garcia-Campos <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0001-5223-8174" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li><a href="authors.html">More about authors...</a></li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/AngelCampos/txtools" class="external-link"><img src="https://img.shields.io/badge/devel%20version-1.0.2-blue.svg"></a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Miguel Angel Garcia-Campos.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
