<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="txtools">
<title>txtools User Guide • txtools</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="txtools User Guide">
<meta property="og:description" content="txtools">
<meta property="og:image" content="https://angelcampos.github.io/txtools/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">txtools</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.5</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/txtools.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>txtools User Guide</h1>
                        <h4 data-toc-skip class="author">Miguel Angel
Garcia-Campos</h4>
            
      
      
      <div class="d-none name"><code>txtools.Rmd</code></div>
    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>R package that enables the processing, analysis, and
      visualization of RNA-seq data at the nucleotide-level resolution,
      seamlessly integrating alignments to the genome with
      transcriptomic representation.</p>
    </div>
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The following is a quick but in-depth tutorial showing all of txtools
‘core’ functions.</p>
<p>We will show how to:</p>
<ol style="list-style-type: decimal">
<li>Load read-alignments data, and the genomic and transcriptomic
reference files into R.</li>
<li>Assign alignments to genes.</li>
<li>Summarize mapped-reads into different kinds of tables.</li>
<li>Use tx_add_* functions to add useful data and calculated metrics to
the summarized tables.</li>
<li>Plot regions and analyses results using the tx_plot_* functions
family.</li>
<li>Use tx_get_*() functions to extract non-standard information
structures from txtools tables.</li>
<li>Use tx_test_*() functions to apply statiscal testing to detect
transcriptomics positions that have unexpected values at either
continuous or count data variables.</li>
</ol>
<p>The aim of txtools is to provide a versatile framework to analyze
transcriptomic data in a simple manner. To accomplish this we propose
the processing of data in two stages:</p>
<ol style="list-style-type: decimal">
<li>Transfer read-alignments from a <strong>genomic</strong> to a
<strong>transcriptomic</strong> coordinates space.</li>
<li>Summarize transcriptomic-alignments into table structures.</li>
</ol>
<div class="float">
<img src="../reference/figures/readme_1.png" alt="Fig 1. txtools main processing workflow"><div class="figcaption">Fig 1. txtools main processing workflow</div>
</div>
<p>The intermediate output in Step 1 is of class <code>GRanges</code>
from the <em><a href="https://bioconductor.org/packages/3.13/GenomicRanges" class="external-link">GenomicRanges</a></em>
package, while that of Step 2, and final output of the main txtools
processing pipeline, is of class <code>data.table</code> from the <em><a href="https://CRAN.R-project.org/package=data.table" class="external-link">data.table</a></em>
package, which is an extension of the <code>data.frame</code> class.</p>
<blockquote>
<p>Current txtools development has favored functionality on the more
common table structure, as it offers extendability and versatility but
at the price of big memory allocation. This should be considered when
developing analysis pipelines using txtools. Users will benefit benefit
of not using the whole transcriptome at early development stages, unless
necessary, and instead use a subset of the gene annotation while
developing the analysis pipeline and only at late stages run over the
complete transcriptome.</p>
</blockquote>
<p><em>For this example we will use the data of the </em><a href="https://bioconductor.org/packages/3.13/pasillaBamSubset" class="external-link">pasillaBamSubset</a><em>
package. A subset of reads aligned to the chromosome 4 of D.
melanogaster.</em></p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://doi.org/10.1093/nar/gkae203" class="external-link">txtools</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">pasillaBamSubset</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="loading-data-tx_load_">Loading data <code>tx_load_*()</code><a class="anchor" aria-label="anchor" href="#loading-data-tx_load_"></a>
</h2>
<p>To work with the reads in the transcriptomic space, txtools requires
a <strong>Gene Annotation</strong> file in the form of a BED-12 or BED-6
file, loaded with <code><a href="../reference/tx_load_bed.html">tx_load_bed()</a></code>.</p>
<p><em>Importantly, a BED-12 file contains the information</em> <em>to
construct the exon structure, when using a BED-6 file, txtools will
assume</em> <em>that the genes in the annotation consist of a single
exon block.</em></p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">BED_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tx_dm3_geneAnnot.html">tx_dm3_geneAnnot</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># geneAnnot built-in txtools</span></span>
<span><span class="va">geneAnnot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tx_load_bed.html">tx_load_bed</a></span><span class="op">(</span><span class="va">BED_path</span><span class="op">)</span></span></code></pre></div>
<p>Additionally, it is recommended to have a <strong>Reference
Genome</strong> sequence file in the form of a FASTA file, in which each
entry corresponds to a chromosome, loaded with
<code><a href="../reference/tx_load_genome.html">tx_load_genome()</a></code>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">FASTA_path</span> <span class="op">&lt;-</span> <span class="fu">pasillaBamSubset</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/pasillaBamSubset/man/pasillaBamSubset-package.html" class="external-link">dm3_chr4</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># D. melanogaster chr4 built-in pasillaBamSubset</span></span>
<span><span class="va">genome</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tx_load_genome.html">tx_load_genome</a></span><span class="op">(</span>fastaFile <span class="op">=</span> <span class="va">FASTA_path</span><span class="op">)</span></span></code></pre></div>
<p>The main data that we will work with and process using txtools are
read mappings, or read alignments, in the form of <strong>BAM
files</strong>. These files can be loaded into R using
<code><a href="../reference/tx_load_bam.html">tx_load_bam()</a></code>. In this case we will set the
<code>pairedEnd</code> argument to TRUE, to specify that the library is
paired-end; as we will use the sequence information from the alignments
we need to set the <code>loadSeq</code> argument also to TRUE.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a>BAM_path <span class="ot">&lt;-</span> pasillaBamSubset<span class="sc">::</span><span class="fu">untreated3_chr4</span>() <span class="co"># Paired-end BAM file built-in pasillaBamSubset</span></span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a>genAligns <span class="ot">&lt;-</span> <span class="fu">tx_load_bam</span>(<span class="at">file =</span> BAM_path, <span class="at">pairedEnd =</span> <span class="cn">TRUE</span>, <span class="at">loadSeq =</span> <span class="cn">TRUE</span>, <span class="at">loadSecondaryAligns =</span> <span class="cn">FALSE</span>)</span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a><span class="co">#&gt; Reading number of records in file </span></span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a><span class="co">#&gt; 175346 number of BAM records </span></span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a><span class="co">#&gt; Loading BAM file </span></span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a>  <span class="er">|</span>                                                                      <span class="er">|</span>   <span class="dv">0</span>%</span>
<span id="cb4-9"><a href="#cb4-9" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb4-10"><a href="#cb4-10" tabindex="-1"></a>  <span class="er">|======================================================================|</span> <span class="dv">100</span>%</span>
<span id="cb4-11"><a href="#cb4-11" tabindex="-1"></a><span class="co">#&gt; 0 reads filtered out for empty sequence field </span></span>
<span id="cb4-12"><a href="#cb4-12" tabindex="-1"></a><span class="co">#&gt; 75409 Reads succesfully loaded </span></span>
<span id="cb4-13"><a href="#cb4-13" tabindex="-1"></a><span class="co">#&gt; Dumped ambiguous reads: 0</span></span></code></pre></div>
<p>These three functions make loading the necessary input data a
painless process.</p>
</div>
<div class="section level2">
<h2 id="assigning-alignments-to-genes-tx_reads">Assigning alignments to genes <code>tx_reads()</code><a class="anchor" aria-label="anchor" href="#assigning-alignments-to-genes-tx_reads"></a>
</h2>
<p>Reads mapped to a reference genome have to be assigned and processed
to their genes. This is done using <code><a href="../reference/tx_reads.html">tx_reads()</a></code>.</p>
<p><code><a href="../reference/tx_reads.html">tx_reads()</a></code> accepts bam files loaded with
<code><a href="../reference/tx_load_bam.html">tx_load_bam()</a></code> recognizing if they are paired-end or
single-end libraries. Additionally it is required to provide a
gene-annotation as loaded with <code><a href="../reference/tx_load_bed.html">tx_load_bed()</a></code>.</p>
<p><em>By default <code><a href="../reference/tx_reads.html">tx_reads()</a></code> will only process genes with
50 or more reads,</em> <em>will not output the processed sequence for
each read, and will use only</em> <em>one core; all these options can be
changed in its arguments.</em></p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">txReads</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tx_reads.html">tx_reads</a></span><span class="op">(</span><span class="va">genAligns</span>, geneAnnot <span class="op">=</span> <span class="va">geneAnnot</span>, minReads <span class="op">=</span> <span class="fl">1</span>, withSeq <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in rmAmbigStrandAligns(reads): Removing 63 alignments with ambiguous strand (*)</span></span>
<span><span class="co">#&gt; Processing 75346 reads, using 10 gene models. </span></span>
<span><span class="co">#&gt; 10373 alignments overlap 10 gene models </span></span>
<span><span class="co">#&gt; Assigning alignments to gene model... </span></span>
<span><span class="co">#&gt; Processing sequences. This may take several minutes depending on geneAnnot size ... </span></span>
<span><span class="co">#&gt; Output contains: 7176 unique alignments in 10 gene models</span></span>
<span><span class="co">#&gt; Warning in tx_reads(genAligns, geneAnnot = geneAnnot, minReads = 1, withSeq =</span></span>
<span><span class="co">#&gt; TRUE): Some alignments were not assigned to any gene, you can retrieve them</span></span>
<span><span class="co">#&gt; using the tx_get_unassignedAlignments() function.</span></span></code></pre></div>
<p>To retrieve the genomic reads that where not assigned to any gene,
due to being outside of the gene annotation, you can use
<code><a href="../reference/tx_get_unassignedAlignments.html">tx_getUnassignedAlignments()</a></code>. <em>Background on the
<code>GenomicRanges</code> package is necessary for further</em>
<em>examination and manipulation of not assigned reads</em></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/tx_get_unassignedAlignments.html">tx_getUnassignedAlignments</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: `tx_getUnassignedAlignments()` was deprecated in txtools 1.0.0.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Please use `tx_get_unassignedAlignments()` instead.</span></span>
<span><span class="co">#&gt; <span style="color: #555555;">This warning is displayed once every 8 hours.</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;">Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;">generated.</span></span></span>
<span><span class="co">#&gt; GAlignmentPairs object with 68170 pairs, strandMode=1, and 0 metadata columns:</span></span>
<span><span class="co">#&gt;                     seqnames strand   :          ranges  --          ranges</span></span>
<span><span class="co">#&gt;                        &lt;Rle&gt;  &lt;Rle&gt;   :       &lt;IRanges&gt;  --       &lt;IRanges&gt;</span></span>
<span><span class="co">#&gt;   SRR031715.1138209     chr4      +   :         169-205  --         326-362</span></span>
<span><span class="co">#&gt;    SRR031714.756385     chr4      +   :         943-979  --       1086-1122</span></span>
<span><span class="co">#&gt;   SRR031714.2355189     chr4      +   :         944-980  --       1119-1155</span></span>
<span><span class="co">#&gt;   SRR031714.5054563     chr4      +   :         946-982  --        986-1022</span></span>
<span><span class="co">#&gt;   SRR031715.1722593     chr4      +   :        966-1002  --       1108-1144</span></span>
<span><span class="co">#&gt;                 ...      ...    ... ...             ... ...             ...</span></span>
<span><span class="co">#&gt;   SRR031714.2561111     chr4      -   : 1348209-1348245  -- 1348063-1348099</span></span>
<span><span class="co">#&gt;   SRR031715.2322156     chr4      -   : 1348211-1348247  -- 1348077-1348113</span></span>
<span><span class="co">#&gt;   SRR031715.2887514     chr4      +   : 1348217-1348253  -- 1348215-1348251</span></span>
<span><span class="co">#&gt;   SRR031714.1650928     chr4      +   : 1349196-1349232  -- 1349326-1349362</span></span>
<span><span class="co">#&gt;   SRR031714.1650928     chr4      +   : 1349708-1349744  -- 1349838-1349874</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   seqinfo: 8 sequences from an unspecified genome</span></span></code></pre></div>
<p>In this toy example most reads are not mapped as we only used a gene
annotation with ten genes for assigning reads.</p>
<p>The resulting object <code>txReads</code> comprises a list of
GRanges, genomic ranges by name, but which consist of the ranges of the
mapped reads in the transcriptomic space, not in genomic space.</p>
</div>
<div class="section level2">
<h2 id="summarizing-reads-into-a-txdt-tx_makedt_">Summarizing reads into a txDT <code>tx_makeDT_*()</code><a class="anchor" aria-label="anchor" href="#summarizing-reads-into-a-txdt-tx_makedt_"></a>
</h2>
<p>After consolidating the paired-end reads, assigning them to their
respective genes and converting their ranges into transcriptomic space,
we can summarize their data into table format. These tables contain data
such as the coverage, and nucleotide frequency counts; each of the rows
in the table corresponds to a nucleotide position unique in the
transcriptome, but could have duplicated genomic coordinate, as is the
case for overlapping genes.</p>
<p>By default R base uses the class object <code>data.frame</code> to
accomodate table data structures which contain different classes of data
per column. For the case of txtools we have opted to use the class
object <a href="https://cran.r-project.org/web/packages/data.table/index.html" class="external-link"><code>data.table</code></a>
from the eponymous package, as the data.table object operates very
similar to base-R’s data.frames yet are optimized for speedier
calculations and assignment.</p>
<p>There are three functions to generate summarized txDTs, each
containing summarized data per nucleotide position in the
transcriptome:</p>
<ul>
<li>
<code><a href="../reference/tx_makeDT_coverage.html">tx_makeDT_coverage()</a></code>: Generates a table with coverage,
read-starts, and read-ends counts data.</li>
</ul>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DT_cov</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tx_makeDT_coverage.html">tx_makeDT_coverage</a></span><span class="op">(</span><span class="va">txReads</span>, geneAnnot <span class="op">=</span> <span class="va">geneAnnot</span><span class="op">)</span></span>
<span><span class="va">DT_cov</span></span>
<span><span class="co">#&gt;         chr gencoor strand         gene txcoor cov start_5p end_3p</span></span>
<span><span class="co">#&gt;     1: chr4  150475      - NM_001169348      1   0        0      0</span></span>
<span><span class="co">#&gt;     2: chr4  150474      - NM_001169348      2   0        0      0</span></span>
<span><span class="co">#&gt;     3: chr4  150473      - NM_001169348      3   0        0      0</span></span>
<span><span class="co">#&gt;     4: chr4  150472      - NM_001169348      4   0        0      0</span></span>
<span><span class="co">#&gt;     5: chr4  150471      - NM_001169348      5   0        0      0</span></span>
<span><span class="co">#&gt;    ---                                                            </span></span>
<span><span class="co">#&gt; 42061: chr4 1133504      - NM_001297793   6073   0        0      0</span></span>
<span><span class="co">#&gt; 42062: chr4 1133503      - NM_001297793   6074   0        0      0</span></span>
<span><span class="co">#&gt; 42063: chr4 1133502      - NM_001297793   6075   0        0      0</span></span>
<span><span class="co">#&gt; 42064: chr4 1133501      - NM_001297793   6076   0        0      0</span></span>
<span><span class="co">#&gt; 42065: chr4 1133500      - NM_001297793   6077   0        0      0</span></span></code></pre></div>
<ul>
<li>
<code><a href="../reference/tx_makeDT_nucFreq.html">tx_makeDT_nucFreq()</a></code>: Generates a table with nucleotide
frequencies for each of the four main aminoacids, unidentified
nucleotide (N), deletions (-), and insert counts (.). The latter are
generated when using paired-end reads that didn’t overlap with each
other, and have an unidentified sequence between Read1 and Read2.</li>
</ul>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DT_nucFreq</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tx_makeDT_nucFreq.html">tx_makeDT_nucFreq</a></span><span class="op">(</span><span class="va">txReads</span>, geneAnnot <span class="op">=</span> <span class="va">geneAnnot</span><span class="op">)</span></span>
<span><span class="va">DT_nucFreq</span></span>
<span><span class="co">#&gt;         chr gencoor strand         gene txcoor A C G T N - .</span></span>
<span><span class="co">#&gt;     1: chr4  150475      - NM_001169348      1 0 0 0 0 0 0 0</span></span>
<span><span class="co">#&gt;     2: chr4  150474      - NM_001169348      2 0 0 0 0 0 0 0</span></span>
<span><span class="co">#&gt;     3: chr4  150473      - NM_001169348      3 0 0 0 0 0 0 0</span></span>
<span><span class="co">#&gt;     4: chr4  150472      - NM_001169348      4 0 0 0 0 0 0 0</span></span>
<span><span class="co">#&gt;     5: chr4  150471      - NM_001169348      5 0 0 0 0 0 0 0</span></span>
<span><span class="co">#&gt;    ---                                                      </span></span>
<span><span class="co">#&gt; 42061: chr4 1133504      - NM_001297793   6073 0 0 0 0 0 0 0</span></span>
<span><span class="co">#&gt; 42062: chr4 1133503      - NM_001297793   6074 0 0 0 0 0 0 0</span></span>
<span><span class="co">#&gt; 42063: chr4 1133502      - NM_001297793   6075 0 0 0 0 0 0 0</span></span>
<span><span class="co">#&gt; 42064: chr4 1133501      - NM_001297793   6076 0 0 0 0 0 0 0</span></span>
<span><span class="co">#&gt; 42065: chr4 1133500      - NM_001297793   6077 0 0 0 0 0 0 0</span></span></code></pre></div>
<ul>
<li>
<code><a href="../reference/tx_makeDT_covNucFreq.html">tx_makeDT_covNucFreq()</a></code>: Generates a table with all
columns from the two previous functions.</li>
</ul>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DT_covNucFreq</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tx_makeDT_covNucFreq.html">tx_makeDT_covNucFreq</a></span><span class="op">(</span><span class="va">txReads</span>, geneAnnot <span class="op">=</span> <span class="va">geneAnnot</span><span class="op">)</span> </span></code></pre></div>
<p>The three of these tables contain 5 columns that identify each
transcriptomic position and link it to their respective genomic position
as well:</p>
<ul>
<li>chr: Chromosome</li>
<li>gencoor: Genomic coordinate</li>
<li>strand: Genomic strand</li>
<li>gene: Gene name</li>
<li>txcoor: Transcriptomic coordinate, or relative coordinate in the
transcript.</li>
</ul>
<p><em><code>txcoor</code> includes UTR regions if present, i.e. the
first nucleotide of the 5’UTR is </em> <em>considered to be position
“1”.</em></p>
<p>Additionally, in each of the functions when the <code>genome</code>
argument is input with its corresponding reference genome sequence, the
nucleotide identity per position is added in the column
<strong><code>refSeq</code></strong>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DT_covNucFreq</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tx_makeDT_covNucFreq.html">tx_makeDT_covNucFreq</a></span><span class="op">(</span><span class="va">txReads</span>, geneAnnot <span class="op">=</span> <span class="va">geneAnnot</span>, genome <span class="op">=</span> <span class="va">genome</span><span class="op">)</span> </span>
<span><span class="va">DT_covNucFreq</span></span>
<span><span class="co">#&gt;         chr gencoor strand         gene txcoor refSeq cov start_5p end_3p A C G</span></span>
<span><span class="co">#&gt;     1: chr4  150475      - NM_001169348      1      C   0        0      0 0 0 0</span></span>
<span><span class="co">#&gt;     2: chr4  150474      - NM_001169348      2      T   0        0      0 0 0 0</span></span>
<span><span class="co">#&gt;     3: chr4  150473      - NM_001169348      3      T   0        0      0 0 0 0</span></span>
<span><span class="co">#&gt;     4: chr4  150472      - NM_001169348      4      T   0        0      0 0 0 0</span></span>
<span><span class="co">#&gt;     5: chr4  150471      - NM_001169348      5      A   0        0      0 0 0 0</span></span>
<span><span class="co">#&gt;    ---                                                                         </span></span>
<span><span class="co">#&gt; 42061: chr4 1133504      - NM_001297793   6073      A   0        0      0 0 0 0</span></span>
<span><span class="co">#&gt; 42062: chr4 1133503      - NM_001297793   6074      A   0        0      0 0 0 0</span></span>
<span><span class="co">#&gt; 42063: chr4 1133502      - NM_001297793   6075      T   0        0      0 0 0 0</span></span>
<span><span class="co">#&gt; 42064: chr4 1133501      - NM_001297793   6076      T   0        0      0 0 0 0</span></span>
<span><span class="co">#&gt; 42065: chr4 1133500      - NM_001297793   6077      C   0        0      0 0 0 0</span></span>
<span><span class="co">#&gt;        T N - .</span></span>
<span><span class="co">#&gt;     1: 0 0 0 0</span></span>
<span><span class="co">#&gt;     2: 0 0 0 0</span></span>
<span><span class="co">#&gt;     3: 0 0 0 0</span></span>
<span><span class="co">#&gt;     4: 0 0 0 0</span></span>
<span><span class="co">#&gt;     5: 0 0 0 0</span></span>
<span><span class="co">#&gt;    ---        </span></span>
<span><span class="co">#&gt; 42061: 0 0 0 0</span></span>
<span><span class="co">#&gt; 42062: 0 0 0 0</span></span>
<span><span class="co">#&gt; 42063: 0 0 0 0</span></span>
<span><span class="co">#&gt; 42064: 0 0 0 0</span></span>
<span><span class="co">#&gt; 42065: 0 0 0 0</span></span></code></pre></div>
<ul>
<li>
<strong>Note:</strong> The reason for which summarized tables
shouldn’t be generated in genomic<em> </em>space is that complex genes,
as those from human, have very long introns that <em> </em>would
generate too long and sparse tables, yet too heavy in terms of memory
allocation.*</li>
</ul>
<hr>
</div>
<div class="section level2">
<h2 id="adding-useful-data-and-metrics-to-txdts-tx_add_">Adding useful data and metrics to txDTs <code>tx_add_*()</code><a class="anchor" aria-label="anchor" href="#adding-useful-data-and-metrics-to-txdts-tx_add_"></a>
</h2>
<p>After generating a txDT we already have all the information we need
to do most of our count data analysis. Yet, sometimes we need this data
as other metrics or ratios, consider for example the ratio of
read-starts compared to coverage, which we call the read-start ratio or
more simply startRatio.</p>
<p>We could manually calculate the startRatio by dividing the
<code>start_5p</code> column over the <code>cov</code> column of our
txDT. Yet other metrics in which the limits of the genes are important
would be off calculated, e.g. startRatio 1bp upstream. To easily
calculate different metrics and add additional information to a txDT we
provide a set of functions that start with the <code>tx_add_</code>
prefix.</p>
<div class="section level3">
<h3 id="adding-metrics-related-to-read-starts-read-ends-coverage-and-nucleotide-frequency-">Adding metrics related to read-starts, read-ends, coverage, and
nucleotide frequency.<a class="anchor" aria-label="anchor" href="#adding-metrics-related-to-read-starts-read-ends-coverage-and-nucleotide-frequency-"></a>
</h3>
<p>Additional to the counts of coverage (cov), read-starts (start_5p),
read-ends (end_3p), and nucleotide frequencies (A, T, C, G) for each
nucleotide in the transcriptome, it can be useful to add certain related
metrics.</p>
<p>For example the rate at which reads start compared to coverage, which
we call startRatio. To create such common metrics we provide several
functions under the <code>tx_add_*()</code> functions family. Bellow a
list of them and a brief description of what they add.</p>
<ul>
<li>
<code><a href="../reference/tx_add_startRatio.html">tx_add_startRatio()</a></code>: Start to coverage rate</li>
<li>
<code><a href="../reference/tx_add_startRatio1bpDS.html">tx_add_startRatio1bpDS()</a></code>: Start to coverage rate 1 bp
down-stream</li>
<li>
<code><a href="../reference/tx_add_startRatio1bpUS.html">tx_add_startRatio1bpUS()</a></code>: Start to coverage rate 1 bp
up-stream</li>
<li>
<code><a href="../reference/tx_add_endRatio.html">tx_add_endRatio()</a></code>: End to coverage rate</li>
<li>
<code><a href="../reference/tx_add_endRatio1bpDS.html">tx_add_endRatio1bpDS()</a></code>: End to coverage rate 1 bp
down-stream</li>
<li>
<code><a href="../reference/tx_add_endRatio1bpUS.html">tx_add_endRatio1bpUS()</a></code>: End to coverage rate 1 bp
up-stream</li>
<li>
<code><a href="../reference/tx_add_nucTotal.html">tx_add_nucTotal()</a></code>: Sum of nucleotide frequencies, not
counting undetermined ‘N’ and inserts ‘.’.</li>
<li>
<code><a href="../reference/tx_add_misincCount.html">tx_add_misincCount()</a></code>: Sum of nucleotide reads different
to the reference sequence.</li>
<li>
<code><a href="../reference/tx_add_misincRate.html">tx_add_misincRate()</a></code>: Rate of nucleotide reads different
to the reference sequence.</li>
<li>
<code><a href="../reference/tx_add_misincRateNucSpec.html">tx_add_misincRateNucSpec()</a></code>: Rate of misincorporation of
an specific nucleotide with respect to a reference nucleotide. e.g. the
rate at which ‘T’ was incorporated (or misread) in the RNA instead of a
‘C’.</li>
<li>
<code><a href="../reference/tx_add_geneRegion.html">tx_add_geneRegion()</a></code>: Adds the gene region to which the
positions correspond: 5’UTR, CDS, 3’UTR, or non-coding.</li>
</ul>
<p><em>The result of using a <code>tx_add_()</code> function is the same
txDT input with the</em> <em>difference of a new column of the desired
metric, hence it is possible to</em> <em>chain tx_add functions in
succession with the pipe operator <code>%&gt;%</code> from the</em> **<a href="https://CRAN.R-project.org/package=magrittr" class="external-link">magrittr</a>*
package.*</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://magrittr.tidyverse.org" class="external-link">"magrittr"</a></span><span class="op">)</span></span>
<span><span class="va">DT_covNucFreq</span> <span class="op">&lt;-</span> <span class="va">DT_covNucFreq</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="../reference/tx_add_startRatio.html">tx_add_startRatio</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="../reference/tx_add_startRatio1bpDS.html">tx_add_startRatio1bpDS</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="../reference/tx_add_startRatio1bpUS.html">tx_add_startRatio1bpUS</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="../reference/tx_add_endRatio.html">tx_add_endRatio</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="../reference/tx_add_endRatio1bpDS.html">tx_add_endRatio1bpDS</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="../reference/tx_add_endRatio1bpUS.html">tx_add_endRatio1bpUS</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="../reference/tx_add_nucTotal.html">tx_add_nucTotal</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="../reference/tx_add_misincCount.html">tx_add_misincCount</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="../reference/tx_add_misincRate.html">tx_add_misincRate</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="../reference/tx_add_misincRateNucSpec.html">tx_add_misincRateNucSpec</a></span><span class="op">(</span>refNuc <span class="op">=</span> <span class="st">"C"</span>, misNuc <span class="op">=</span> <span class="st">"T"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="../reference/tx_add_geneRegion.html">tx_add_geneRegion</a></span><span class="op">(</span>geneAnnot <span class="op">=</span> <span class="va">geneAnnot</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="adding-unique-position-identifier">Adding unique position identifier<a class="anchor" aria-label="anchor" href="#adding-unique-position-identifier"></a>
</h3>
<p>If needed, by combining the gene and relative coordinate we can
create a single unique identifier for each position in the
transcriptome. We can add this with <code><a href="../reference/tx_add_pos.html">tx_add_pos()</a></code>.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DT_covNucFreq</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tx_add_pos.html">tx_add_pos</a></span><span class="op">(</span><span class="va">DT_covNucFreq</span><span class="op">)</span></span></code></pre></div>
<p><em>IMPORTANT NOTE: Due to being unique identifiers the
<code>pos</code> column is of</em> <em>character class and it requires a
big amount of memory to be allocated. If not </em> <em>needed is
recommended to omit the <code>pos</code> column or delete it after no
longer is</em> <em>needed.</em></p>
</div>
<div class="section level3">
<h3 id="adding-reference-sequence">Adding reference sequence<a class="anchor" aria-label="anchor" href="#adding-reference-sequence"></a>
</h3>
<p>If the genome sequence was not specified in the tx_makeDT_*()
function call, adding the transcriptomic reference sequence after
generating a txDT is possible with <code><a href="../reference/tx_add_refSeqDT.html">tx_add_refSeqDT()</a></code>.</p>
<p>Importantly, the sequence in refSeq has already been processed to
represent the transcriptomic sequence considering the strand of the
gene.</p>
<p><em>Note: Importantly, the nucleotide uracil is represented with the
letter “T”</em> <em>instead of with the letter “U”. This might change in
future versions if </em> <em>poses a conflict or confusion to the
users.</em></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DT_covNucFreq</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tx_add_refSeqDT.html">tx_add_refSeqDT</a></span><span class="op">(</span><span class="va">DT_covNucFreq</span>, genome <span class="op">=</span> <span class="va">genome</span>, geneAnnot <span class="op">=</span> <span class="va">geneAnnot</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="adding-sites-annotation">Adding sites annotation<a class="anchor" aria-label="anchor" href="#adding-sites-annotation"></a>
</h3>
<p>Sometimes we want to identify previously annotated positions of a
transcriptome in our data. For this, we can use
<code><a href="../reference/tx_add_siteAnnotation.html">tx_add_siteAnnotation()</a></code> to add a logical vector that
indicates the sites of interest to perform further analysis and
plotting.</p>
<p>To add a sites annotation we only need to input a GRanges object that
consists of ranges of length 1, which can also be loaded with
<code><a href="../reference/tx_load_bed.html">tx_load_bed()</a></code>.</p>
<p><em>Here we use the in-built site annotation object
<code>annotSites_RRACH</code> which</em> <em>contains 636 sites that
fall in the arbitrarily selected RRACH motif.</em></p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DT_covNucFreq</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tx_add_siteAnnotation.html">tx_add_siteAnnotation</a></span><span class="op">(</span><span class="va">DT_covNucFreq</span>, GRanges <span class="op">=</span> <span class="va">annotSites_RRACH</span>, colName <span class="op">=</span> <span class="st">"annotSites"</span><span class="op">)</span></span></code></pre></div>
<p><em>We will visualize the effect of annotating this RRACH sites with
a seqlogo</em> <em>plot in section @ref(ggseqlogo-plot)</em></p>
</div>
<div class="section level3">
<h3 id="adding-sequence-motif-presence">Adding sequence motif presence<a class="anchor" aria-label="anchor" href="#adding-sequence-motif-presence"></a>
</h3>
<p>To mark all the positions at which specific sequences appear on
transcripts in the txDT we can use
<code><a href="../reference/tx_add_motifPresence.html">tx_add_motifPresence()</a></code>.</p>
<p><em>Importantly, this function will take a character string that can
consist of</em> <em>the characters in the <a href="https://en.wikipedia.org/wiki/Nucleic_acid_notation" class="external-link">IUPAC
nucleotide ambiguity code</a></em></p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DT_covNucFreq</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tx_add_motifPresence.html">tx_add_motifPresence</a></span><span class="op">(</span><span class="va">DT_covNucFreq</span>, motif <span class="op">=</span> <span class="st">"DRACH"</span>, nucPositions <span class="op">=</span> <span class="st">"center"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="adding-rolling-mean">Adding rolling mean<a class="anchor" aria-label="anchor" href="#adding-rolling-mean"></a>
</h3>
<p>Sometime we need a smoother version of a metric, which can be
acomplished by computing a roll mean across a moving window. For example
we can smooth the coverage by applying a rolling mean across the
<code>cov</code> column. The <code>winSize</code> argument determines
the length of the window over which the mean is calculatates and the
<code>align</code>.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DT_covNucFreq</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tx_add_rollingMean.html">tx_add_rollingMean</a></span><span class="op">(</span><span class="va">DT_covNucFreq</span>, colName <span class="op">=</span> <span class="st">"cov"</span>, winSize <span class="op">=</span> <span class="fl">50</span>, align <span class="op">=</span> <span class="st">"left"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="plotting-functions-tx_plot_">Plotting functions <code>tx_plot_*()</code><a class="anchor" aria-label="anchor" href="#plotting-functions-tx_plot_"></a>
</h2>
<p>txtools provides a diverse set of functions to plot. For ease of use
all of them start with the prefix <code>tx_plot_</code></p>
<div class="section level3">
<h3 id="start-end-coverage-plot">Start End Coverage plot<a class="anchor" aria-label="anchor" href="#start-end-coverage-plot"></a>
</h3>
<p><code><a href="../reference/tx_plot_staEndCov.html">tx_plot_staEndCov()</a></code>: This plot is useful to visualize
the coverage, read-start, and read-end patterns. The required input is a
txDT, the gene, and the transcriptomic range to be plot.</p>
<p>It can be used to plot the whole gene as bellow:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/tx_plot_staEndCov.html">tx_plot_staEndCov</a></span><span class="op">(</span><span class="va">DT_covNucFreq</span>, gene <span class="op">=</span> <span class="st">"NM_001272152"</span>, show_yLabels <span class="op">=</span> <span class="cn">FALSE</span>, bar_border <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="txtools_files/figure-html/unnamed-chunk-18-1.png" width="576"></p>
<p>Or part of the gene specifying the relative coordinates range in the
<code>txRange</code> argument.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/tx_plot_staEndCov.html">tx_plot_staEndCov</a></span><span class="op">(</span><span class="va">DT_covNucFreq</span>, gene <span class="op">=</span> <span class="st">"NM_001272152"</span>, txRange <span class="op">=</span> <span class="fl">330</span><span class="op">:</span><span class="fl">360</span><span class="op">)</span></span></code></pre></div>
<p><img src="txtools_files/figure-html/unnamed-chunk-19-1.png" width="576"></p>
</div>
<div class="section level3">
<h3 id="nucleotide-frequency-plot">Nucleotide frequency plot<a class="anchor" aria-label="anchor" href="#nucleotide-frequency-plot"></a>
</h3>
<p><code><a href="../reference/tx_plot_nucFreq.html">tx_plot_nucFreq()</a></code>: This plot shows the nucleotide
frequency and importantly, shows in color the nucleotides that are not
matching the reference sequence. It is useful to visualize
misincorporation rate, which can be originated due to a SNP, chemical
treatment on RNA or (less informative) sequencing errors.</p>
<blockquote>
<p>Importantly, this plot show the coverage due to inserts, which has
not been assigned a nucleotide identity, as is the unsequenced gap
between read1 and read2 in paired-end libraries</p>
</blockquote>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">selSite</span> <span class="op">&lt;-</span> <span class="va">DT_covNucFreq</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">misincRate</span> <span class="op">&gt;</span> <span class="fl">0.5</span> <span class="op">&amp;</span> <span class="va">nucTotal</span> <span class="op">&gt;</span> <span class="fl">40</span><span class="op">)</span>,</span>
<span>                         <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gene"</span>, <span class="st">"txcoor"</span>, <span class="st">"cov"</span>, <span class="st">"nucTotal"</span>, <span class="st">"misincCount"</span>, <span class="st">"misincRate"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">selSite</span></span>
<span><span class="co">#&gt;         gene txcoor cov nucTotal misincCount misincRate</span></span>
<span><span class="co">#&gt; 1: NM_079901   3803  90       46          46          1</span></span></code></pre></div>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/tx_plot_nucFreq.html">tx_plot_nucFreq</a></span><span class="op">(</span><span class="va">DT_covNucFreq</span>, gene <span class="op">=</span> <span class="va">selSite</span><span class="op">$</span><span class="va">gene</span>, txRange <span class="op">=</span> <span class="fu"><a href="../reference/window_around.html">window_around</a></span><span class="op">(</span><span class="va">selSite</span><span class="op">$</span><span class="va">txcoor</span>, <span class="fl">20</span><span class="op">)</span>, removeInsert <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code></pre></div>
<p><img src="txtools_files/figure-html/unnamed-chunk-21-1.png" width="576"></p>
</div>
<div class="section level3">
<h3 id="ggseqlogo-plot">ggseqlogo plot<a class="anchor" aria-label="anchor" href="#ggseqlogo-plot"></a>
</h3>
<p><code><a href="../reference/tx_plot_ggseqlogo.html">tx_plot_ggseqlogo()</a></code>: This function allows to quickly
plot a <em><a href="https://CRAN.R-project.org/package=ggseqlogo" class="external-link">ggseqlogo</a></em>
using as input a txDT and a logical annotation which marks the center
position to be plot. It is useful to identify motif sequences in sites
that have been selected, for example for passing certain thresholds of
specific metrics.</p>
<p><em>In this example we use the sites we annotated while using
<code><a href="../reference/tx_add_motifPresence.html">tx_add_motifPresence()</a></code></em> <em>for annotating the DRACH
motifs in our small transcriptome set.</em></p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/tx_plot_ggseqlogo.html">tx_plot_ggseqlogo</a></span><span class="op">(</span><span class="va">DT_covNucFreq</span>, logi_col <span class="op">=</span> <span class="st">"DRACH_motif_center"</span>, upFlank <span class="op">=</span> <span class="fl">5</span>, doFlank <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in tx_get_flankSequence(DT = DT, logi_col = logi_col, upFlank =</span></span>
<span><span class="co">#&gt; upFlank, : Some sequences reached the end of transcript, a '.' was added in</span></span>
<span><span class="co">#&gt; place, which may affect downstream results.</span></span>
<span><span class="co">#&gt; Warning: The `&lt;scale&gt;` argument of `guides()` cannot be `FALSE`. Use "none" instead as</span></span>
<span><span class="co">#&gt; of ggplot2 3.3.4.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> The deprecated feature was likely used in the <span style="color: #0000BB;">ggseqlogo</span> package.</span></span>
<span><span class="co">#&gt;   Please report the issue at <span style="color: #0000BB; font-style: italic;">&lt;https://github.com/omarwagih/ggseqlogo/issues&gt;</span>.</span></span>
<span><span class="co">#&gt; <span style="color: #555555;">This warning is displayed once every 8 hours.</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;">Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;">generated.</span></span></span></code></pre></div>
<p><img src="txtools_files/figure-html/unnamed-chunk-22-1.png" width="576"></p>
</div>
<div class="section level3">
<h3 id="metagene-plot-at-cds-startend">Metagene plot at CDS start/end<a class="anchor" aria-label="anchor" href="#metagene-plot-at-cds-startend"></a>
</h3>
<p><code><a href="../reference/tx_plot_metageneAtCDS.html">tx_plot_metageneAtCDS()</a></code>: Calculates and plots metagenes
using as input a txDT and the column names of the metrics to be
plot.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/tx_plot_metageneAtCDS.html">tx_plot_metageneAtCDS</a></span><span class="op">(</span>txDT <span class="op">=</span> <span class="va">DT_covNucFreq</span>, geneAnnot <span class="op">=</span> <span class="va">geneAnnot</span>,</span>
<span>                      colVars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"G"</span>, <span class="st">"C"</span>, <span class="st">"T"</span><span class="op">)</span>, CDS_align <span class="op">=</span> <span class="st">"spliceSite"</span>, </span>
<span>                      upFlank <span class="op">=</span> <span class="fl">200</span>, doFlank <span class="op">=</span> <span class="fl">200</span><span class="op">)</span></span></code></pre></div>
<p><img src="txtools_files/figure-html/unnamed-chunk-23-1.png" width="576"></p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/tx_plot_metageneAtCDS.html">tx_plot_metageneAtCDS</a></span><span class="op">(</span>txDT <span class="op">=</span> <span class="va">DT_covNucFreq</span>, geneAnnot <span class="op">=</span> <span class="va">geneAnnot</span>,</span>
<span>                      colVars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cov"</span>, <span class="st">"nucTotal"</span><span class="op">)</span>, CDS_align <span class="op">=</span> <span class="st">"spliceSite"</span>, </span>
<span>                      upFlank <span class="op">=</span> <span class="fl">100</span>, doFlank <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span></code></pre></div>
<p><img src="txtools_files/figure-html/unnamed-chunk-23-2.png" width="576"></p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/tx_plot_metageneExons.html">tx_plot_metageneExons</a></span><span class="op">(</span>txDT <span class="op">=</span> <span class="va">DT_covNucFreq</span>, colVars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"T"</span>, <span class="st">"G"</span>, <span class="st">"C"</span><span class="op">)</span>, nBins <span class="op">=</span> <span class="fl">30</span>, geneAnnot <span class="op">=</span> <span class="va">geneAnnot</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in tx_get_metageneExons(txDT = txDT, colVars = colVars, nBins = nBins,</span></span>
<span><span class="co">#&gt; : exonNumber is needed in txDT to extract exons, this function calculates it</span></span>
<span><span class="co">#&gt; but discards it everytime. If you want to keep them please add it to the txDT</span></span>
<span><span class="co">#&gt; by using tx_add_exonNumber()</span></span></code></pre></div>
<p><img src="txtools_files/figure-html/unnamed-chunk-24-1.png" width="576"></p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/tx_plot_metageneRegions.html">tx_plot_metageneRegions</a></span><span class="op">(</span>txDT <span class="op">=</span> <span class="va">DT_covNucFreq</span>, colVars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cov"</span>, <span class="st">"nucTotal"</span><span class="op">)</span>,</span>
<span>                        geneAnnot <span class="op">=</span> <span class="va">geneAnnot</span>, nBins_5UTR <span class="op">=</span> <span class="fl">40</span>, nBins_CDS <span class="op">=</span> <span class="fl">80</span>,</span>
<span>                        nBins_3UTR <span class="op">=</span> <span class="fl">80</span>, summ_fun <span class="op">=</span> <span class="st">"mean"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Running using 1 core.</span></span>
<span><span class="co">#&gt;  Consider increasing the  'nCores' argument if function takes too long</span></span></code></pre></div>
<p><img src="txtools_files/figure-html/unnamed-chunk-25-1.png" width="576"></p>
</div>
</div>
<div class="section level2">
<h2 id="get-functions-tx_get_">Get functions <code>tx_get_*()</code><a class="anchor" aria-label="anchor" href="#get-functions-tx_get_"></a>
</h2>
<div class="section level3">
<h3 id="get-flanks-from-annotation">Get flanks from annotation<a class="anchor" aria-label="anchor" href="#get-flanks-from-annotation"></a>
</h3>
<p><code><a href="../reference/tx_get_flanksFromLogicAnnot.html">tx_get_flanksFromLogicAnnot()</a></code>: Extracts the surrounding
values of annotates sites from any column. The required input is the
txDT containing the column from were the values will be extracted alogn
a logical vector marking the annotated sites (TRUE = site of interest,
FALSE = not annotated site). The output is a matrix in which rows
correspond to annoted sites, and columns to the values at each of the
positions up-stream, center, and down-stream from the site of
interest.</p>
<p>Getting this values may be useful to confirm results derived from
previous analysis, or to generate an aggregated analysis of the sites of
interest. Bellow is an example in which we extract the coverage of the
sites annotated in the <code>DRACH_motif_center</code> column.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/tx_get_flanksFromLogicAnnot.html">tx_get_flanksFromLogicAnnot</a></span><span class="op">(</span><span class="va">DT_covNucFreq</span>, logi_col <span class="op">=</span> <span class="st">"DRACH_motif_center"</span>, </span>
<span>                            values_col <span class="op">=</span> <span class="st">"cov"</span>, upFlank <span class="op">=</span> <span class="fl">4</span>, doFlank <span class="op">=</span> <span class="fl">4</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">5</span><span class="op">)</span> </span>
<span><span class="co">#&gt;                  -4 bp -3 bp -2 bp -1 bp DRACH_motif_center 1 bp 2 bp 3 bp 4 bp</span></span>
<span><span class="co">#&gt; NM_001169348:97     10    10    10    10                 10   12   12   12   12</span></span>
<span><span class="co">#&gt; NM_001169348:108    12    12    12    12                 12   13   13   13   13</span></span>
<span><span class="co">#&gt; NM_001169348:120    15    15    15    15                 15   15   15   15   16</span></span>
<span><span class="co">#&gt; NM_001169348:148    17    17    17    17                 17   17   17   17   17</span></span>
<span><span class="co">#&gt; NM_001169348:170    17    17    17    17                 17   17   17   17   18</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="getting-flanking-sequences">Getting flanking sequences<a class="anchor" aria-label="anchor" href="#getting-flanking-sequences"></a>
</h3>
<p><code><a href="../reference/tx_get_flankSequence.html">tx_get_flankSequence()</a></code>: Extracts the contiguous sequence
of sites of interest. Works similarly to
<code><a href="../reference/tx_get_flanksFromLogicAnnot.html">tx_get_flanksFromLogicAnnot()</a></code> but it is a specific case for
getting only the sequences in a collapsed format. Its output is a
character vector in which each element is the sequence at the site of
interest and its upstream and downstream flanking regions.</p>
<p>Bellow is an example showing the flanking sequence of the sites
annotated in the <code>DRACH_morif_center</code> column.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/tx_get_flankSequence.html">tx_get_flankSequence</a></span><span class="op">(</span><span class="va">DT_covNucFreq</span>, logi_col <span class="op">=</span> <span class="st">"DRACH_motif_center"</span>, </span>
<span>                     upFlank <span class="op">=</span> <span class="fl">4</span>, doFlank <span class="op">=</span> <span class="fl">4</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">5</span><span class="op">)</span> </span>
<span><span class="co">#&gt; Warning in tx_get_flankSequence(DT_covNucFreq, logi_col = "DRACH_motif_center",</span></span>
<span><span class="co">#&gt; : Some sequences reached the end of transcript, a '.' was added in place, which</span></span>
<span><span class="co">#&gt; may affect downstream results.</span></span>
<span><span class="co">#&gt; [1] "GAGAACAAA" "ACAAACAAG" "TGAAACCTG" "TTAGACAAC" "TAAAACAGA"</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="get-gene-lengths">Get gene lengths<a class="anchor" aria-label="anchor" href="#get-gene-lengths"></a>
</h3>
<p><code>tx_get_geneLengths</code>: Gets the length of genes contained
in a txDT, as measured by the number of positions (rows) the gene has in
the table.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/tx_get_geneLengths.html">tx_get_geneLengths</a></span><span class="op">(</span><span class="va">DT_covNucFreq</span><span class="op">)</span></span>
<span><span class="co">#&gt; NM_001169348 NM_001272124 NM_001272152 NM_001297793    NM_079901    NM_143656 </span></span>
<span><span class="co">#&gt;         5777         3965         2026         6077         5576         2557 </span></span>
<span><span class="co">#&gt;    NM_143665    NM_166736    NM_166777    NM_166788 </span></span>
<span><span class="co">#&gt;         5357         3050         4068         3612</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="get-metagene-values">Get metagene values<a class="anchor" aria-label="anchor" href="#get-metagene-values"></a>
</h3>
<p><code><a href="../reference/tx_get_metageneAtCDS.html">tx_get_metageneAtCDS()</a></code>: Gets the metagene profile of
desired column variables, centered at either the start or end of the
CDS.</p>
<p>Bellow is an example in which we retrieve the values of coverage at
the start of the CDS as well as upstream and downstream flanks.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/tx_get_metageneAtCDS.html">tx_get_metageneAtCDS</a></span><span class="op">(</span>txDT <span class="op">=</span> <span class="va">DT_covNucFreq</span>, geneAnnot <span class="op">=</span> <span class="va">geneAnnot</span>,</span>
<span>                      colVars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cov"</span>, <span class="st">"nucTotal"</span><span class="op">)</span>, CDS_align <span class="op">=</span> <span class="st">"start"</span>, </span>
<span>                      upFlank <span class="op">=</span> <span class="fl">3</span>, doFlank <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="co">#&gt; $cov</span></span>
<span><span class="co">#&gt;                  -3 bp -2 bp -1 bp CDS_start 1 bp 2 bp 3 bp 4 bp 5 bp</span></span>
<span><span class="co">#&gt; NM_001169348:69      3     3     3         3    3    3    3    3    4</span></span>
<span><span class="co">#&gt; NM_166736:419        4     4     4         4    4    4    4    4    4</span></span>
<span><span class="co">#&gt; NM_143656:186        0     0     0         0    0    0    0    0    0</span></span>
<span><span class="co">#&gt; NM_001272124:60      2     2     2         2    2    2    2    2    2</span></span>
<span><span class="co">#&gt; NM_143665:519       17    17    17        17   17   17   17   17   17</span></span>
<span><span class="co">#&gt; NM_166777:357        1     1     1         1    1    1    1    1    1</span></span>
<span><span class="co">#&gt; NM_166788:511        7     7     7         7    7    7    7    7    7</span></span>
<span><span class="co">#&gt; NM_079901:381       65    65    65        66   66   68   69   68   68</span></span>
<span><span class="co">#&gt; NM_001272152:347    75    76    77        77   77  112  115  115  118</span></span>
<span><span class="co">#&gt; NM_001297793:188     3     3     3         3    3    3    3    3    3</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $nucTotal</span></span>
<span><span class="co">#&gt;                  -3 bp -2 bp -1 bp CDS_start 1 bp 2 bp 3 bp 4 bp 5 bp</span></span>
<span><span class="co">#&gt; NM_001169348:69      3     3     3         3    3    3    3    3    4</span></span>
<span><span class="co">#&gt; NM_166736:419        2     2     2         2    2    2    2    2    2</span></span>
<span><span class="co">#&gt; NM_143656:186        0     0     0         0    0    0    0    0    0</span></span>
<span><span class="co">#&gt; NM_001272124:60      2     2     2         2    2    2    2    2    2</span></span>
<span><span class="co">#&gt; NM_143665:519        5     5     5         6    6    6    7    7    7</span></span>
<span><span class="co">#&gt; NM_166777:357        1     1     1         1    1    1    1    1    1</span></span>
<span><span class="co">#&gt; NM_166788:511        5     5     5         5    5    5    5    5    5</span></span>
<span><span class="co">#&gt; NM_079901:381       11    11    11        12   12   14   15   14   14</span></span>
<span><span class="co">#&gt; NM_001272152:347    75    76    77        77   77  111  114  114  117</span></span>
<span><span class="co">#&gt; NM_001297793:188     3     3     3         3    3    2    2    2    2</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="get-transcriptome-sequences">Get transcriptome sequences<a class="anchor" aria-label="anchor" href="#get-transcriptome-sequences"></a>
</h3>
<p><code><a href="../reference/tx_get_transcriptSeqs.html">tx_get_transcriptSeqs()</a></code>: Retrieves the spliced sequence
of genes. Using as input the reference genome and the gene annotation
loaded with <code><a href="../reference/tx_load_bed.html">tx_load_bed()</a></code>.</p>
<p>The output of this function can be either a DNAStringSet from the
<em><a href="https://bioconductor.org/packages/3.13/Biostrings" class="external-link">Biostrings</a></em>
package, or (when providing a file name) a FASTA file.</p>
<p>Bellow an example showing the sequences of the genes in the D.
melanogaster gene annotation, and the DNAStringSet as output.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/tx_get_transcriptSeqs.html">tx_get_transcriptSeqs</a></span><span class="op">(</span>genome <span class="op">=</span> <span class="va">genome</span>, geneAnnot <span class="op">=</span> <span class="va">geneAnnot</span><span class="op">)</span></span>
<span><span class="co">#&gt; DNAStringSet object of length 10:</span></span>
<span><span class="co">#&gt;      width seq                                              names               </span></span>
<span><span class="co">#&gt;  [1]  5777 <span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFD7AF;">TTT</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span>...<span style="color: #000000; background-color: #AFFFAF;">CCCC</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">CCC</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">CCC</span><span style="color: #000000; background-color: #AFFFFF;">GG</span> NM_001169348</span></span>
<span><span class="co">#&gt;  [2]  3050 <span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFD7AF;">TTTT</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">TT</span>...<span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">AAA</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">AAAA</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #FFAFAF;">AAA</span><span style="color: #000000; background-color: #FFD7AF;">T</span> NM_166736</span></span>
<span><span class="co">#&gt;  [3]  2557 <span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFD7AF;">TTT</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">GGG</span>...<span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">TTT</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">CC</span> NM_143656</span></span>
<span><span class="co">#&gt;  [4]  3965 <span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">TTTT</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #FFD7AF;">TTT</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">TTT</span>...<span style="color: #000000; background-color: #FFAFAF;">AAA</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">TTT</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">AAAAAAA</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span> NM_001272124</span></span>
<span><span class="co">#&gt;  [5]  5357 <span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">TTT</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #FFD7AF;">TTT</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span>...<span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">AAA</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span> NM_143665</span></span>
<span><span class="co">#&gt;  [6]  4068 <span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">AAAA</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span>...<span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">AAA</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span> NM_166777</span></span>
<span><span class="co">#&gt;  [7]  3612 <span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFD7AF;">TTT</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span>...<span style="color: #000000; background-color: #FFAFAF;">AAA</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #AFFFAF;">CCCCCC</span> NM_166788</span></span>
<span><span class="co">#&gt;  [8]  5576 <span style="color: #000000; background-color: #FFAFAF;">AAA</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">AA</span>...<span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">AAAA</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">AAA</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #AFFFAF;">CC</span> NM_079901</span></span>
<span><span class="co">#&gt;  [9]  2026 <span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">TTTTTT</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">TTTT</span><span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFAFAF;">AA</span>...<span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">TTT</span><span style="color: #000000; background-color: #FFAFAF;">AAA</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span> NM_001272152</span></span>
<span><span class="co">#&gt; [10]  6077 <span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span>...<span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #AFFFAF;">C</span> NM_001297793</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="statistical-tests-tx_test_">Statistical tests <code>tx_test_*()</code><a class="anchor" aria-label="anchor" href="#statistical-tests-tx_test_"></a>
</h2>
<p>txtools provides two functions to perform statistical tests to detect
sites that show unexpected values at each site in a treatment group
compared to a control group.</p>
<ul>
<li><p>t test: In the case of comparing continuous variables a t-test
can show the unlikeliness that both groups have the same mean. e.g. whe
comparing start to coverage rate.</p></li>
<li><p>Likelihood Ratio test (from <em><a href="https://bioconductor.org/packages/3.13/edgeR" class="external-link">edgeR</a></em>
package): For comparing count data that are of the type of trial and
success between groups, for example when comparing start to coverage
ratio but directly with the count data, considering coverage counts as
number of trials and the number of read-starts as successes.</p></li>
</ul>
<p>Both of these functions require as main input a list of txDTs
(txDTL), each txDT representing a sample on the experimental design.</p>
<p><em>For examples on the usage of the tx_test functions, please
consult the case</em> <em>studies vignette available at: [PENDING
URL]</em></p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<pre><code><span><span class="co">#&gt; R version 4.1.0 (2021-05-18)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: CentOS Linux 7 (Core)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /apps/RH7U2/gnu/R/4.1.0/lib64/R/lib/libRblas.so</span></span>
<span><span class="co">#&gt; LAPACK: /apps/RH7U2/gnu/R/4.1.0/lib64/R/lib/libRlapack.so</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] pasillaBamSubset_0.34.0 BiocStyle_2.28.0        txtools_1.0.5          </span></span>
<span><span class="co">#&gt; [4] testthat_3.1.4          devtools_2.4.3          usethis_2.1.6          </span></span>
<span><span class="co">#&gt; [7] magrittr_2.0.3         </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] colorspace_2.1-0            rjson_0.2.21               </span></span>
<span><span class="co">#&gt;   [3] ellipsis_0.3.2              rprojroot_2.0.3            </span></span>
<span><span class="co">#&gt;   [5] XVector_0.34.0              GenomicRanges_1.48.0       </span></span>
<span><span class="co">#&gt;   [7] fs_1.6.3                    rstudioapi_0.13            </span></span>
<span><span class="co">#&gt;   [9] farver_2.1.1                remotes_2.4.2              </span></span>
<span><span class="co">#&gt;  [11] bit64_4.0.5                 AnnotationDbi_1.56.2       </span></span>
<span><span class="co">#&gt;  [13] fansi_1.0.4                 splines_4.1.0              </span></span>
<span><span class="co">#&gt;  [15] cachem_1.0.8                knitr_1.44                 </span></span>
<span><span class="co">#&gt;  [17] RcppRoll_0.3.0              pkgload_1.3.0              </span></span>
<span><span class="co">#&gt;  [19] jsonlite_1.8.7              Rsamtools_2.8.0            </span></span>
<span><span class="co">#&gt;  [21] annotate_1.72.0             png_0.1-8                  </span></span>
<span><span class="co">#&gt;  [23] BiocManager_1.30.22         compiler_4.1.0             </span></span>
<span><span class="co">#&gt;  [25] httr_1.4.7                  Matrix_1.4-1               </span></span>
<span><span class="co">#&gt;  [27] fastmap_1.1.1               limma_3.56.2               </span></span>
<span><span class="co">#&gt;  [29] cli_3.6.1                   htmltools_0.5.6            </span></span>
<span><span class="co">#&gt;  [31] prettyunits_1.1.1           tools_4.1.0                </span></span>
<span><span class="co">#&gt;  [33] gtable_0.3.4                glue_1.6.2                 </span></span>
<span><span class="co">#&gt;  [35] GenomeInfoDbData_1.2.7      dplyr_1.1.3                </span></span>
<span><span class="co">#&gt;  [37] Rcpp_1.0.11                 Biobase_2.54.0             </span></span>
<span><span class="co">#&gt;  [39] jquerylib_0.1.4             pkgdown_2.0.7              </span></span>
<span><span class="co">#&gt;  [41] vctrs_0.6.3                 Biostrings_2.62.0          </span></span>
<span><span class="co">#&gt;  [43] rtracklayer_1.54.0          ggseqlogo_0.1              </span></span>
<span><span class="co">#&gt;  [45] xfun_0.40                   stringr_1.5.0              </span></span>
<span><span class="co">#&gt;  [47] plyranges_1.20.0            ps_1.7.1                   </span></span>
<span><span class="co">#&gt;  [49] brio_1.1.3                  lifecycle_1.0.3            </span></span>
<span><span class="co">#&gt;  [51] restfulr_0.0.15             XML_3.99-0.6               </span></span>
<span><span class="co">#&gt;  [53] edgeR_3.42.4                zlibbioc_1.40.0            </span></span>
<span><span class="co">#&gt;  [55] scales_1.2.1                ragg_1.2.5                 </span></span>
<span><span class="co">#&gt;  [57] MatrixGenerics_1.6.0        parallel_4.1.0             </span></span>
<span><span class="co">#&gt;  [59] SummarizedExperiment_1.24.0 yaml_2.3.7                 </span></span>
<span><span class="co">#&gt;  [61] memoise_2.0.1               ggplot2_3.4.3              </span></span>
<span><span class="co">#&gt;  [63] sass_0.4.7                  stringi_1.7.12             </span></span>
<span><span class="co">#&gt;  [65] RSQLite_2.3.1               genefilter_1.74.1          </span></span>
<span><span class="co">#&gt;  [67] S4Vectors_0.38.1            BiocIO_1.4.0               </span></span>
<span><span class="co">#&gt;  [69] desc_1.4.1                  BiocGenerics_0.40.0        </span></span>
<span><span class="co">#&gt;  [71] pkgbuild_1.3.1              BiocParallel_1.28.3        </span></span>
<span><span class="co">#&gt;  [73] GenomeInfoDb_1.36.2         rlang_1.1.1                </span></span>
<span><span class="co">#&gt;  [75] pkgconfig_2.0.3             systemfonts_1.0.4          </span></span>
<span><span class="co">#&gt;  [77] matrixStats_1.0.0           bitops_1.0-7               </span></span>
<span><span class="co">#&gt;  [79] evaluate_0.22               lattice_0.20-45            </span></span>
<span><span class="co">#&gt;  [81] purrr_1.0.2                 labeling_0.4.3             </span></span>
<span><span class="co">#&gt;  [83] GenomicAlignments_1.36.0    bit_4.0.5                  </span></span>
<span><span class="co">#&gt;  [85] processx_3.7.0              tidyselect_1.2.0           </span></span>
<span><span class="co">#&gt;  [87] bookdown_0.35               R6_2.5.1                   </span></span>
<span><span class="co">#&gt;  [89] IRanges_2.34.1              generics_0.1.3             </span></span>
<span><span class="co">#&gt;  [91] DelayedArray_0.20.0         DBI_1.1.3                  </span></span>
<span><span class="co">#&gt;  [93] pillar_1.9.0                withr_2.5.1                </span></span>
<span><span class="co">#&gt;  [95] survival_3.3-1              KEGGREST_1.34.0            </span></span>
<span><span class="co">#&gt;  [97] RCurl_1.98-1.6              tibble_3.2.1               </span></span>
<span><span class="co">#&gt;  [99] crayon_1.5.2                utf8_1.2.3                 </span></span>
<span><span class="co">#&gt; [101] rmarkdown_2.25              locfit_1.5-9.8             </span></span>
<span><span class="co">#&gt; [103] grid_4.1.0                  data.table_1.14.8          </span></span>
<span><span class="co">#&gt; [105] blob_1.2.4                  callr_3.7.3                </span></span>
<span><span class="co">#&gt; [107] forcats_1.0.0               digest_0.6.33              </span></span>
<span><span class="co">#&gt; [109] xtable_1.8-4                tidyr_1.3.0                </span></span>
<span><span class="co">#&gt; [111] textshaping_0.3.6           stats4_4.1.0               </span></span>
<span><span class="co">#&gt; [113] munsell_0.5.0               bslib_0.5.1                </span></span>
<span><span class="co">#&gt; [115] sessioninfo_1.2.2</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Miguel Angel Garcia-Campos.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
