<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Function reference â€¢ txtools</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">txtools</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.2</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/txtools.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Function reference</h1>
    </div>

    <div class="section level2">
      <h2 id="all-functions">All functions<a class="anchor" aria-label="anchor" href="#all-functions"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="hid_aggregate_DTlist.html">hid_aggregate_DTlist()</a></code> 
        </dt>
        <dd>Merge lists of data.tables</dd>
      </dl><dl><dt>
          
          <code><a href="hlp_coverageTab_mc.html">hlp_coverageTab_mc()</a></code> 
        </dt>
        <dd>Calculate coverage table (coverage, ends, starts) for all gene models
without gene coordinates (Multicore)</dd>
      </dl><dl><dt>
          
          <code><a href="hlp_genCoorTab_mc.html">hlp_genCoorTab_mc()</a></code> 
        </dt>
        <dd>Coordinate system table generator</dd>
      </dl><dl><dt>
          
          <code><a href="hlp_nucFreqTab_mc.html">hlp_nucFreqTab_mc()</a></code> 
        </dt>
        <dd>Calculate nucleotide frequency pileup for all gene models</dd>
      </dl><dl><dt>
          
          <code><a href="IUPAC_code_2nucs.html">IUPAC_code_2nucs</a></code> 
        </dt>
        <dd>IUPAC ambiguity alphabet (2nc)</dd>
      </dl><dl><dt>
          
          <code><a href="IUPAC_CODE_MAP_extended.html">IUPAC_CODE_MAP_extended</a></code> 
        </dt>
        <dd>IUPAC code extended to consider deletions</dd>
      </dl><dl><dt>
          
          <code><a href="IUPAC_code_simpl.html">IUPAC_code_simpl</a></code> 
        </dt>
        <dd>Simplified nucleotide alphabet</dd>
      </dl><dl><dt>
          
          <code><a href="txtools.html">txtools</a></code> <code><a href="txtools.html">txtools-package</a></code> 
        </dt>
        <dd>txtools: A package facilitating analysis of RNA modifications, structures, and interactions</dd>
      </dl><dl><dt>
          
          <code><a href="tx_add_endRatio.html">tx_add_endRatio()</a></code> 
        </dt>
        <dd>Add ends to coverage ratio 1 bp downstream</dd>
      </dl><dl><dt>
          
          <code><a href="tx_add_motifPresence.html">tx_add_motifPresence()</a></code> 
        </dt>
        <dd>Add motif presence</dd>
      </dl><dl><dt>
          
          <code><a href="tx_add_nucTotal.html">tx_add_nucTotal()</a></code> 
        </dt>
        <dd>Add total number of nucleotide reads</dd>
      </dl><dl><dt>
          
          <code><a href="tx_add_pos.html">tx_add_pos()</a></code> 
        </dt>
        <dd>Add position unique names</dd>
      </dl><dl><dt>
          
          <code><a href="tx_add_refSeqDT.html">tx_add_refSeqDT()</a></code> 
        </dt>
        <dd>Add reference sequence to a data.table</dd>
      </dl><dl><dt>
          
          <code><a href="tx_add_siteAnnotation.html">tx_add_siteAnnotation()</a></code> 
        </dt>
        <dd>Add 1bp-site logical annotation</dd>
      </dl><dl><dt>
          
          <code><a href="tx_add_startRatio.html">tx_add_startRatio()</a></code> 
        </dt>
        <dd>Add starts to coverage ratio</dd>
      </dl><dl><dt>
          
          <code><a href="tx_complete_DT.html">tx_complete_DT()</a></code> 
        </dt>
        <dd>Complete a DT object missing genes</dd>
      </dl><dl><dt>
          
          <code><a href="tx_counts.html">tx_counts()</a></code> 
        </dt>
        <dd>Total counts of reads per gene model</dd>
      </dl><dl><dt>
          
          <code><a href="tx_cutEnds_DT.html">tx_cutEnds_DT()</a></code> 
        </dt>
        <dd>Cutting 5' and 3' ends of data.table using txcoors</dd>
      </dl><dl><dt>
          
          <code><a href="tx_dm3_geneAnnot.html">tx_dm3_geneAnnot()</a></code> 
        </dt>
        <dd>D. melanogaster gene annotation subset path</dd>
      </dl><dl><dt>
          
          <code><a href="tx_extend_UTR.html">tx_extend_UTR()</a></code> 
        </dt>
        <dd>Extending GRanges 5' and 3' UTR blocks</dd>
      </dl><dl><dt>
          
          <code><a href="tx_filter_maxWidth.html">tx_filter_maxWidth()</a></code> 
        </dt>
        <dd>Filter ranges by a maximum width</dd>
      </dl><dl><dt>
          
          <code><a href="tx_load_bam.html">tx_load_bam()</a></code> 
        </dt>
        <dd>Read paired end bam file by yield size</dd>
      </dl><dl><dt>
          
          <code><a href="tx_load_bed.html">tx_load_bed()</a></code> 
        </dt>
        <dd>Load gene models from bed-12 and bed-6 files</dd>
      </dl><dl><dt>
          
          <code><a href="tx_load_genome.html">tx_load_genome()</a></code> 
        </dt>
        <dd>Load genome</dd>
      </dl><dl><dt>
          
          <code><a href="tx_makeDT_coverage.html">tx_makeDT_coverage()</a></code> 
        </dt>
        <dd>Summarized Coverage data.table</dd>
      </dl><dl><dt>
          
          <code><a href="tx_makeDT_covNucFreq.html">tx_makeDT_covNucFreq()</a></code> 
        </dt>
        <dd>Summarized Coverage &amp; Nucleotide Frequency data.table</dd>
      </dl><dl><dt>
          
          <code><a href="tx_makeDT_nucFreq.html">tx_makeDT_nucFreq()</a></code> 
        </dt>
        <dd>Summarized Nucleotide Frequency data.table</dd>
      </dl><dl><dt>
          
          <code><a href="tx_merge_DT.html">tx_merge_DT()</a></code> 
        </dt>
        <dd>Merge data.tables in list to a single data.table</dd>
      </dl><dl><dt>
          
          <code><a href="tx_plot_nucFreq.html">tx_plot_nucFreq()</a></code> 
        </dt>
        <dd>Nucleotide frequency plot</dd>
      </dl><dl><dt>
          
          <code><a href="tx_plot_staEndCov.html">tx_plot_staEndCov()</a></code> 
        </dt>
        <dd>Transcript coverage plot highlighting read-starts and read-ends counts</dd>
      </dl><dl><dt>
          
          <code><a href="tx_reads.html">tx_reads()</a></code> 
        </dt>
        <dd>Transcriptomic reads convertion</dd>
      </dl><dl><dt>
          
          <code><a href="tx_sample_GRList.html">tx_sample_GRList()</a></code> 
        </dt>
        <dd>Sampling alignments</dd>
      </dl><dl><dt>
          
          <code><a href="tx_split_DT.html">tx_split_DT()</a></code> 
        </dt>
        <dd>Split data.table to list of data.tables</dd>
      </dl><dl><dt>
          
          <code><a href="window_around.html">window_around()</a></code> 
        </dt>
        <dd>Centered numeric sequence</dd>
      </dl><dl><dt>
          
          <code><a href="annotSites_RRACH.html">annotSites_RRACH</a></code> 
        </dt>
        <dd>RRACH sites annotation D. melanogaster</dd>
      </dl><dl><dt>
          
          <code><a href="bam_sk1.html">bam_sk1</a></code> 
        </dt>
        <dd>Genomic Alignments example - Sk1 yeast</dd>
      </dl><dl><dt>
          
          <code><a href="gA_sk1.html">gA_sk1</a></code> 
        </dt>
        <dd>Gene annotation example - Sk1 yeast</dd>
      </dl><dl><dt>
          
          <code><a href="genome_sk1.html">genome_sk1</a></code> 
        </dt>
        <dd>Genome example - Sk1 yeast</dd>
      </dl><dl><dt>
          
          <code><a href="getDumpedAlignments.html">getDumpedAlignments()</a></code> 
        </dt>
        <dd>Retrieve dumped alignments</dd>
      </dl><dl><dt>
          
          <code><a href="mm_geneAnnot.html">mm_geneAnnot()</a></code> 
        </dt>
        <dd>Download gene annotation - Case study # 2</dd>
      </dl><dl><dt>
          
          <code><a href="sc_geneAnnot.html">sc_geneAnnot()</a></code> 
        </dt>
        <dd>Download gene annotation - Case study #1</dd>
      </dl><dl><dt>
          
          <code><a href="sc_rRNAmods_Taoka.html">sc_rRNAmods_Taoka</a></code> 
        </dt>
        <dd>Sk1 rRNA modifications catalogue</dd>
      </dl><dl><dt>
          
          <code><a href="sc_txDTL.html">sc_txDTL</a></code> 
        </dt>
        <dd>txtools-processed data - Case study # 1</dd>
      </dl><dl><dt>
          
          <code><a href="tk_faGenome.html">tk_faGenome()</a></code> 
        </dt>
        <dd>Download genome to temporary file - Case study # 3</dd>
      </dl><dl><dt>
          
          <code><a href="tk_geneAnnot.html">tk_geneAnnot()</a></code> 
        </dt>
        <dd>Download gene annotation - Case study # 3</dd>
      </dl><dl><dt>
          
          <code><a href="txDTL_Tk.html">txDTL_Tk</a></code> 
        </dt>
        <dd>txtools-processed data - Case study # 3</dd>
      </dl><dl><dt>
          
          <code><a href="tx_add_endRatio1bpUS.html">tx_add_endRatio1bpUS()</a></code> 
        </dt>
        <dd>Add ends to coverage ratio 1bp up-stream</dd>
      </dl><dl><dt>
          
          <code><a href="tx_add_misincCount.html">tx_add_misincCount()</a></code> 
        </dt>
        <dd>Add counts of nucleotide reads different to the reference sequence</dd>
      </dl><dl><dt>
          
          <code><a href="tx_add_misincRate.html">tx_add_misincRate()</a></code> 
        </dt>
        <dd>Add misincorporation to total nucleotide reads ratio</dd>
      </dl><dl><dt>
          
          <code><a href="tx_add_misincRateNucSpec.html">tx_add_misincRateNucSpec()</a></code> 
        </dt>
        <dd>Nucleotide specific misincorporation rate</dd>
      </dl><dl><dt>
          
          <code><a href="tx_add_rollingMean.html">tx_add_rollingMean()</a></code> 
        </dt>
        <dd>Add rolling mean</dd>
      </dl><dl><dt>
          
          <code><a href="tx_add_startRatio1bpDS.html">tx_add_startRatio1bpDS()</a></code> 
        </dt>
        <dd>Add starts to coverage ratio 1 bp downstream</dd>
      </dl><dl><dt>
          
          <code><a href="tx_data_caseStudy2.html">tx_data_caseStudy2()</a></code> 
        </dt>
        <dd>Load data - Case study #2</dd>
      </dl><dl><dt>
          
          <code><a href="tx_generateSingleEndFASTQ.html">tx_generateSingleEndFASTQ()</a></code> 
        </dt>
        <dd>Generate single-end FASTQ file</dd>
      </dl><dl><dt>
          
          <code><a href="tx_get_flankSequence.html">tx_get_flankSequence()</a></code> 
        </dt>
        <dd>Get flanking sequences</dd>
      </dl><dl><dt>
          
          <code><a href="tx_get_flanksFromLogicAnnot.html">tx_get_flanksFromLogicAnnot()</a></code> 
        </dt>
        <dd>Get data from a position and their neighboring positions</dd>
      </dl><dl><dt>
          
          <code><a href="tx_get_geneLengths.html">tx_get_geneLengths()</a></code> 
        </dt>
        <dd>Get length of genes</dd>
      </dl><dl><dt>
          
          <code><a href="tx_get_metageneAtCDS.html">tx_get_metageneAtCDS()</a></code> 
        </dt>
        <dd>Get metagene at CDS</dd>
      </dl><dl><dt>
          
          <code><a href="tx_get_transcriptSeqs.html">tx_get_transcriptSeqs()</a></code> 
        </dt>
        <dd>Get transcriptome sequences</dd>
      </dl><dl><dt>
          
          <code><a href="tx_load_rdsDT.html">tx_load_rdsDT()</a></code> 
        </dt>
        <dd>Loading RDS files into data.tables</dd>
      </dl><dl><dt>
          
          <code><a href="tx_orderDT.html">tx_orderDT()</a></code> 
        </dt>
        <dd>Order txDT</dd>
      </dl><dl><dt>
          
          <code><a href="tx_plot_ggseqlogo.html">tx_plot_ggseqlogo()</a></code> 
        </dt>
        <dd>Plot motif centered in logical annotation</dd>
      </dl><dl><dt>
          
          <code><a href="tx_plot_metageneAtCDS.html">tx_plot_metageneAtCDS()</a></code> 
        </dt>
        <dd>Plot metagene at CDS</dd>
      </dl><dl><dt>
          
          <code><a href="tx_shift_geneWise.html">tx_shift_geneWise()</a></code> 
        </dt>
        <dd>Shift column in txDT</dd>
      </dl><dl><dt>
          
          <code><a href="tx_test_LRTedgeR.html">tx_test_LRTedgeR()</a></code> 
        </dt>
        <dd>Likelihood Ratio Test</dd>
      </dl><dl><dt>
          
          <code><a href="tx_test_ttest.html">tx_test_ttest()</a></code> 
        </dt>
        <dd>t-test in txDT list</dd>
      </dl><dl><dt>
          
          <code><a href="tx_unifyTxDTL.html">tx_unifyTxDTL()</a></code> 
        </dt>
        <dd>Unify lists of txDTs</dd>
      </dl><dl><dt>
          
          <code><a href="tx_add_spliceSitesLogical.html">tx_add_spliceSitesLogical()</a></code> 
        </dt>
        <dd>Add SpliceSites</dd>
      </dl><dl><dt>
          
          <code><a href="sc_faGenome.html">sc_faGenome()</a></code> 
        </dt>
        <dd>Download genome to temporary file - Case study # 1</dd>
      </dl><dl><dt>
          
          <code><a href="tx_add_endRatio1bpDS.html">tx_add_endRatio1bpDS()</a></code> 
        </dt>
        <dd>Add ends to coverage ratio 1bp down-stream</dd>
      </dl><dl><dt>
          
          <code><a href="tx_add_startRatio1bpUS.html">tx_add_startRatio1bpUS()</a></code> 
        </dt>
        <dd>Add starts to coverage ratio 1 bp upstream</dd>
      </dl><dl><dt>
          
          <code><a href="hlp_plot_metageneRegions.html">hlp_plot_metageneRegions()</a></code> 
        </dt>
        <dd>Plot an extracted metaGeneMatrix</dd>
      </dl><dl><dt>
          
          <code><a href="tx_add_exonNumber.html">tx_add_exonNumber()</a></code> 
        </dt>
        <dd>Add exon number</dd>
      </dl><dl><dt>
          
          <code><a href="tx_add_exonPlace.html">tx_add_exonPlace()</a></code> 
        </dt>
        <dd>Add exon place</dd>
      </dl><dl><dt>
          
          <code><a href="tx_add_geneRegion.html">tx_add_geneRegion()</a></code> 
        </dt>
        <dd>Add gene regions</dd>
      </dl><dl><dt>
          
          <code><a href="tx_add_relTxPos.html">tx_add_relTxPos()</a></code> 
        </dt>
        <dd>Add relative position in transcript</dd>
      </dl><dl><dt>
          
          <code><a href="tx_cut_geneAnnotBytxDT.html">tx_cut_geneAnnotBytxDT()</a></code> 
        </dt>
        <dd>Cut gene annotation by txDT's genes</dd>
      </dl><dl><dt>
          
          <code><a href="tx_get_metageneRegions.html">tx_get_metageneRegions()</a></code> 
        </dt>
        <dd>Get metagene regions</dd>
      </dl><dl><dt>
          
          <code><a href="tx_plot_metageneExons.html">tx_plot_metageneExons()</a></code> 
        </dt>
        <dd>Plot metagene exons</dd>
      </dl><dl><dt>
          
          <code><a href="tx_plot_metageneRegions.html">tx_plot_metageneRegions()</a></code> 
        </dt>
        <dd>Plot metagene by regions</dd>
      </dl><dl><dt>
          
          <code><a href="tx_sampleByGenes.html">tx_sampleByGenes()</a></code> 
        </dt>
        <dd>Sample txDT by genes</dd>
      </dl><dl><dt>
          
          <code><a href="tx_generatePairedEndFASTQ.html">tx_generatePairedEndFASTQ()</a></code> 
        </dt>
        <dd>Generate paired-end FASTQ file</dd>
      </dl><dl><dt>
          
          <code><a href="tx_get_metageneExons.html">tx_get_metageneExons()</a></code> 
        </dt>
        <dd>Get metagene by exons</dd>
      </dl><dl><dt>
          
          <code><a href="tx_get_unassignedAlignments.html">tx_get_unassignedAlignments()</a></code> 
        </dt>
        <dd>Retrieve dumped alignments</dd>
      </dl><dl><dt>
          
          <code><a href="tx_plot_numeric.html">tx_plot_numeric()</a></code> 
        </dt>
        <dd>Numeric plot</dd>
      </dl><dl><dt>
          
          <code><a href="txCoreCols.html">txCoreCols</a></code> 
        </dt>
        <dd>txtools' core cols</dd>
      </dl><dl><dt>
          
          <code><a href="txCoreCols_refSeq.html">txCoreCols_refSeq</a></code> 
        </dt>
        <dd>txtools' core cols and refSeq</dd>
      </dl><dl><dt>
          
          <code><a href="tx_plot_metaGeneByBins.html">tx_plot_metaGeneByBins()</a></code> 
        </dt>
        <dd>Plot metagene by bins</dd>
      </dl></div>
  </main></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Miguel Angel Garcia-Campos.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

