<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="This function constructs a list of data.tables that contains nucleotide frequency
metrics per nucleotide by transcript:
A = Adenine
C = Cytosine
G = Guanine
T = Thymine
N = Undetermined nucleotide
- = Deletion
. = Insert, not read gap between read1 and read2
The function requires the input of a GRangesList object output by the
tx_reads function, which should contain sequence alignments in the
transcriptomic space, and a gene annotation in GRanges format, as loaded by
the tx_load_bed function."><title>Summarized Nucleotide Frequency data.table — tx_makeDT_nucFreq • txtools</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Summarized Nucleotide Frequency data.table — tx_makeDT_nucFreq"><meta property="og:description" content="This function constructs a list of data.tables that contains nucleotide frequency
metrics per nucleotide by transcript:
A = Adenine
C = Cytosine
G = Guanine
T = Thymine
N = Undetermined nucleotide
- = Deletion
. = Insert, not read gap between read1 and read2
The function requires the input of a GRangesList object output by the
tx_reads function, which should contain sequence alignments in the
transcriptomic space, and a gene annotation in GRanges format, as loaded by
the tx_load_bed function."><meta property="og:image" content="https://angelcampos.github.io/txtools/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">txtools</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.3</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/txtools.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Summarized Nucleotide Frequency data.table</h1>
      
      <div class="d-none name"><code>tx_makeDT_nucFreq.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function constructs a list of data.tables that contains nucleotide frequency
metrics per nucleotide by transcript:</p><ul><li><p>A = Adenine</p></li>
<li><p>C = Cytosine</p></li>
<li><p>G = Guanine</p></li>
<li><p>T = Thymine</p></li>
<li><p>N = Undetermined nucleotide</p></li>
<li><p>- = Deletion</p></li>
<li><p>. = Insert, not read gap between read1 and read2</p></li>
</ul><p>The function requires the input of a GRangesList object output by the
<code><a href="tx_reads.html">tx_reads</a></code> function, which should contain sequence alignments in the
transcriptomic space, and a gene annotation in GRanges format, as loaded by
the <code><a href="tx_load_bed.html">tx_load_bed</a></code> function.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">tx_makeDT_nucFreq</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  <span class="va">geneAnnot</span>,</span>
<span>  genome <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  simplify_IUPAC <span class="op">=</span> <span class="st">"splitForceInt"</span>,</span>
<span>  fullDT <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  nCores <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>x</dt>
<dd><p>CompressedGRangesList. Genomic Ranges list containing genomic
alignments data by gene. Constructed via the <code><a href="tx_reads.html">tx_reads</a></code> function.</p></dd>


<dt>geneAnnot</dt>
<dd><p>GRanges. Gene annotation as loaded by <code><a href="tx_load_bed.html">tx_load_bed</a></code>()</p></dd>


<dt>genome</dt>
<dd><p>list. The full reference genome sequences, as loaded by
<code><a href="tx_load_genome.html">tx_load_genome</a></code>() or prepackaged by BSgenome, see ?BSgenome::available.genomes</p></dd>


<dt>simplify_IUPAC</dt>
<dd><p>string. Available options are :</p><ul><li><p>"not": Will output the complete nucleotide frequency table including
ambiguous reads using the IUPAC ambiguity code.
See: <code><a href="https://rdrr.io/pkg/Biostrings/man/IUPAC_CODE_MAP.html" class="external-link">IUPAC_CODE_MAP</a></code></p></li>
<li><p>"splitForceInt" (Default): Will force an integers split in which ambiguous codes
will be split and assigned half the frequency into their respective nucleotides,
if the frequency is an odd number the uneven count will be assigned as "N".</p></li>
<li><p>"splitHalf": Ambiguous nucleotide frequencies will be split in half to
their corresponding nucleotides, in cases where frequency is odd creating
non-integer frequencies.</p></li>
</ul></dd>


<dt>fullDT</dt>
<dd><p>logical. Set to TRUE if it is desired to output a data.table
with all genes and in the same order as 'geneAnnot' object.</p></dd>


<dt>nCores</dt>
<dd><p>integer. Number of cores to run the function with. Multicore
capability not available in Windows OS.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>data.table</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This function allows for usage of multiple cores to reduce processing times
in UNIX-like OS.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p>Other makeDT functions: 
<code><a href="tx_makeDT_covNucFreq.html">tx_makeDT_covNucFreq</a>()</code>,
<code><a href="tx_makeDT_coverage.html">tx_makeDT_coverage</a>()</code></p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>M.A. Garcia-Campos</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Miguel Angel Garcia-Campos.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

